<!DOCTYPE html>
<html lang="en-US" dir="ltr">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Full API | MolecularEvolution.jl</title>
    <meta name="description" content="Documentation for MolecularEvolution.jl">
    <meta name="generator" content="VitePress v1.6.3">
    <link rel="preload stylesheet" href="/MolecularEvolution.jl/previews/PR57/assets/style.J8orhqBf.css" as="style">
    <link rel="preload stylesheet" href="/MolecularEvolution.jl/previews/PR57/vp-icons.css" as="style">
    
    <script type="module" src="/MolecularEvolution.jl/previews/PR57/assets/app.CsH27g6M.js"></script>
    <link rel="preload" href="/MolecularEvolution.jl/previews/PR57/assets/inter-roman-latin.Di8DUHzh.woff2" as="font" type="font/woff2" crossorigin="">
    <link rel="modulepreload" href="/MolecularEvolution.jl/previews/PR57/assets/chunks/theme.xAhEsUIq.js">
    <link rel="modulepreload" href="/MolecularEvolution.jl/previews/PR57/assets/chunks/framework.BPK7t3Yd.js">
    <link rel="modulepreload" href="/MolecularEvolution.jl/previews/PR57/assets/api.md.B_f5Qzf5.lean.js">
    <link rel="icon" href="/favicon.ico">
    <script src="/MolecularEvolution.jl/versions.js"></script>
    <script src="/MolecularEvolution.jl/previews/PR57/siteinfo.js"></script>
    <script id="check-dark-mode">(()=>{const e=localStorage.getItem("vitepress-theme-appearance")||"auto",a=window.matchMedia("(prefers-color-scheme: dark)").matches;(!e||e==="auto"?a:e==="dark")&&document.documentElement.classList.add("dark")})();</script>
    <script id="check-mac-os">document.documentElement.classList.toggle("mac",/Mac|iPhone|iPod|iPad/i.test(navigator.platform));</script>
  </head>
  <body>
    <div id="app"><div class="Layout" data-v-a9a9e638><!--[--><!--]--><!--[--><span tabindex="-1" data-v-492508fc></span><a href="#VPContent" class="VPSkipLink visually-hidden" data-v-492508fc>Skip to content</a><!--]--><!----><header class="VPNav" data-v-a9a9e638 data-v-f1e365da><div class="VPNavBar" data-v-f1e365da data-v-822684d1><div class="wrapper" data-v-822684d1><div class="container" data-v-822684d1><div class="title" data-v-822684d1><div class="VPNavBarTitle has-sidebar" data-v-822684d1 data-v-0f4f798b><a class="title" href="/MolecularEvolution.jl/previews/PR57/" data-v-0f4f798b><!--[--><!--]--><!--[--><img class="VPImage logo" src="/MolecularEvolution.jl/previews/PR57/logo.png" width="24" height="24" alt data-v-35a7d0b8><!--]--><span data-v-0f4f798b>MolecularEvolution.jl</span><!--[--><!--]--></a></div></div><div class="content" data-v-822684d1><div class="content-body" data-v-822684d1><!--[--><!--]--><div class="VPNavBarSearch search" data-v-822684d1><!--[--><!----><div id="local-search"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><span class="vp-icon DocSearch-Search-Icon"></span><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"><kbd class="DocSearch-Button-Key"></kbd><kbd class="DocSearch-Button-Key">K</kbd></span></button></div><!--]--></div><nav aria-labelledby="main-nav-aria-label" class="VPNavBarMenu menu" data-v-822684d1 data-v-e6d46098><span id="main-nav-aria-label" class="visually-hidden" data-v-e6d46098> Main Navigation </span><!--[--><!--[--><div class="VPFlyout VPNavBarMenuGroup" data-v-e6d46098 data-v-04f5c5e9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" data-v-04f5c5e9><span class="text" data-v-04f5c5e9><!----><span data-v-04f5c5e9>Core Concepts</span><span class="vpi-chevron-down text-icon" data-v-04f5c5e9></span></span></button><div class="menu" data-v-04f5c5e9><div class="VPMenu" data-v-04f5c5e9 data-v-7dd3104a><div class="items" data-v-7dd3104a><!--[--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/previews/PR57/intro" data-v-acbfed09><!--[--><span data-v-acbfed09>Intro</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/previews/PR57/framework" data-v-acbfed09><!--[--><span data-v-acbfed09>The MolecularEvolution.jl Framework</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/previews/PR57/models" data-v-acbfed09><!--[--><span data-v-acbfed09>Models</span><!--]--></a></div><!--]--><!--]--></div><!--[--><!--]--></div></div></div><!--]--><!--[--><div class="VPFlyout VPNavBarMenuGroup" data-v-e6d46098 data-v-04f5c5e9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" data-v-04f5c5e9><span class="text" data-v-04f5c5e9><!----><span data-v-04f5c5e9>Methods & Algorithms</span><span class="vpi-chevron-down text-icon" data-v-04f5c5e9></span></span></button><div class="menu" data-v-04f5c5e9><div class="VPMenu" data-v-04f5c5e9 data-v-7dd3104a><div class="items" data-v-7dd3104a><!--[--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/previews/PR57/simulation" data-v-acbfed09><!--[--><span data-v-acbfed09>Simulation</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/previews/PR57/optimization" data-v-acbfed09><!--[--><span data-v-acbfed09>Optimization</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/previews/PR57/ancestors" data-v-acbfed09><!--[--><span data-v-acbfed09>Ancestral Reconstruction</span><!--]--></a></div><!--]--><!--]--></div><!--[--><!--]--></div></div></div><!--]--><!--[--><div class="VPFlyout VPNavBarMenuGroup" data-v-e6d46098 data-v-04f5c5e9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" data-v-04f5c5e9><span class="text" data-v-04f5c5e9><!----><span data-v-04f5c5e9>Extensions & Utilities</span><span class="vpi-chevron-down text-icon" data-v-04f5c5e9></span></span></button><div class="menu" data-v-04f5c5e9><div class="VPMenu" data-v-04f5c5e9 data-v-7dd3104a><div class="items" data-v-7dd3104a><!--[--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/previews/PR57/IO" data-v-acbfed09><!--[--><span data-v-acbfed09>Input/Output</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/previews/PR57/generated/viz" data-v-acbfed09><!--[--><span data-v-acbfed09>Visualization</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/previews/PR57/generated/update" data-v-acbfed09><!--[--><span data-v-acbfed09>Updating a phylogenetic tree</span><!--]--></a></div><!--]--><!--]--></div><!--[--><!--]--></div></div></div><!--]--><!--[--><a class="VPLink link VPNavBarMenuLink" href="/MolecularEvolution.jl/previews/PR57/examples" tabindex="0" data-v-e6d46098 data-v-956ec74c><!--[--><span data-v-956ec74c>Examples</span><!--]--></a><!--]--><!--[--><a class="VPLink link VPNavBarMenuLink active" href="/MolecularEvolution.jl/previews/PR57/api" tabindex="0" data-v-e6d46098 data-v-956ec74c><!--[--><span data-v-956ec74c>Full API</span><!--]--></a><!--]--><!--[--><!----><!--]--><!--]--></nav><!----><div class="VPNavBarAppearance appearance" data-v-822684d1 data-v-af096f4a><button class="VPSwitch VPSwitchAppearance" type="button" role="switch" title aria-checked="false" data-v-af096f4a data-v-e40a8bb6 data-v-4a1c76db><span class="check" data-v-4a1c76db><span class="icon" data-v-4a1c76db><!--[--><span class="vpi-sun sun" data-v-e40a8bb6></span><span class="vpi-moon moon" data-v-e40a8bb6></span><!--]--></span></span></button></div><div class="VPSocialLinks VPNavBarSocialLinks social-links" data-v-822684d1 data-v-164c457f data-v-ee7a9424><!--[--><a class="VPSocialLink no-icon" href="https://github.com/MurrellGroup/MolecularEvolution.jl" aria-label="github" target="_blank" rel="noopener" data-v-ee7a9424 data-v-d26d30cb><span class="vpi-social-github"></span></a><!--]--></div><div class="VPFlyout VPNavBarExtra extra" data-v-822684d1 data-v-925effce data-v-04f5c5e9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" aria-label="extra navigation" data-v-04f5c5e9><span class="vpi-more-horizontal icon" data-v-04f5c5e9></span></button><div class="menu" data-v-04f5c5e9><div class="VPMenu" data-v-04f5c5e9 data-v-7dd3104a><!----><!--[--><!--[--><!----><div class="group" data-v-925effce><div class="item appearance" data-v-925effce><p class="label" data-v-925effce>Appearance</p><div class="appearance-action" data-v-925effce><button class="VPSwitch VPSwitchAppearance" type="button" role="switch" title aria-checked="false" data-v-925effce data-v-e40a8bb6 data-v-4a1c76db><span class="check" data-v-4a1c76db><span class="icon" data-v-4a1c76db><!--[--><span class="vpi-sun sun" data-v-e40a8bb6></span><span class="vpi-moon moon" data-v-e40a8bb6></span><!--]--></span></span></button></div></div></div><div class="group" data-v-925effce><div class="item social-links" data-v-925effce><div class="VPSocialLinks social-links-list" data-v-925effce data-v-ee7a9424><!--[--><a class="VPSocialLink no-icon" href="https://github.com/MurrellGroup/MolecularEvolution.jl" aria-label="github" target="_blank" rel="noopener" data-v-ee7a9424 data-v-d26d30cb><span class="vpi-social-github"></span></a><!--]--></div></div></div><!--]--><!--]--></div></div></div><!--[--><!--[--><!--[--><div class="VPFlyout VPNolebaseEnhancedReadabilitiesMenu VPNolebaseEnhancedReadabilitiesMenuFlyout" aria-label="Enhanced Readability" role="menuitem" data-v-04f5c5e9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" data-v-04f5c5e9><span class="text" data-v-04f5c5e9><span class="i-icon-park-outline:book-open option-icon" data-v-04f5c5e9></span><!----><span class="vpi-chevron-down text-icon" data-v-04f5c5e9></span></span></button><div class="menu" data-v-04f5c5e9><div class="VPMenu" data-v-04f5c5e9 data-v-7dd3104a><!----><!--[--><!--]--></div></div></div><!--]--><!--]--><!--]--><button type="button" class="VPNavBarHamburger hamburger" aria-label="mobile navigation" aria-expanded="false" aria-controls="VPNavScreen" data-v-822684d1 data-v-5dea55bf><span class="container" data-v-5dea55bf><span class="top" data-v-5dea55bf></span><span class="middle" data-v-5dea55bf></span><span class="bottom" data-v-5dea55bf></span></span></button></div></div></div></div><div class="divider" data-v-822684d1><div class="divider-line" data-v-822684d1></div></div></div><!----></header><div class="VPLocalNav has-sidebar empty" data-v-a9a9e638 data-v-070ab83d><div class="container" data-v-070ab83d><button class="menu" aria-expanded="false" aria-controls="VPSidebarNav" data-v-070ab83d><span class="vpi-align-left menu-icon" data-v-070ab83d></span><span class="menu-text" data-v-070ab83d>Menu</span></button><div class="VPLocalNavOutlineDropdown" style="--vp-vh:0px;" data-v-070ab83d data-v-168ddf5d><button data-v-168ddf5d>Return to top</button><!----></div></div></div><aside class="VPSidebar" data-v-a9a9e638 data-v-18756405><div class="curtain" data-v-18756405></div><nav class="nav" id="VPSidebarNav" aria-labelledby="sidebar-aria-label" tabindex="-1" data-v-18756405><span class="visually-hidden" id="sidebar-aria-label" data-v-18756405> Sidebar Navigation </span><!--[--><!--]--><!--[--><div class="no-transition group" data-v-9e426adc><section class="VPSidebarItem level-0 collapsible" data-v-9e426adc data-v-a4b0d9bf><div class="item" role="button" tabindex="0" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><h2 class="text" data-v-a4b0d9bf>Core Concepts</h2><div class="caret" role="button" aria-label="toggle section" tabindex="0" data-v-a4b0d9bf><span class="vpi-chevron-right caret-icon" data-v-a4b0d9bf></span></div></div><div class="items" data-v-a4b0d9bf><!--[--><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/intro" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Intro</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/framework" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>The MolecularEvolution.jl Framework</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/models" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Models</p><!--]--></a><!----></div><!----></div><!--]--></div></section></div><div class="no-transition group" data-v-9e426adc><section class="VPSidebarItem level-0 collapsible" data-v-9e426adc data-v-a4b0d9bf><div class="item" role="button" tabindex="0" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><h2 class="text" data-v-a4b0d9bf>Methods & Algorithms</h2><div class="caret" role="button" aria-label="toggle section" tabindex="0" data-v-a4b0d9bf><span class="vpi-chevron-right caret-icon" data-v-a4b0d9bf></span></div></div><div class="items" data-v-a4b0d9bf><!--[--><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/simulation" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Simulation</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/optimization" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Optimization</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/ancestors" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Ancestral Reconstruction</p><!--]--></a><!----></div><!----></div><!--]--></div></section></div><div class="no-transition group" data-v-9e426adc><section class="VPSidebarItem level-0 collapsible" data-v-9e426adc data-v-a4b0d9bf><div class="item" role="button" tabindex="0" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><h2 class="text" data-v-a4b0d9bf>Extensions & Utilities</h2><div class="caret" role="button" aria-label="toggle section" tabindex="0" data-v-a4b0d9bf><span class="vpi-chevron-right caret-icon" data-v-a4b0d9bf></span></div></div><div class="items" data-v-a4b0d9bf><!--[--><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/IO" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Input/Output</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/generated/viz" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Visualization</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/generated/update" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Updating a phylogenetic tree</p><!--]--></a><!----></div><!----></div><!--]--></div></section></div><div class="no-transition group" data-v-9e426adc><section class="VPSidebarItem level-0 has-active" data-v-9e426adc data-v-a4b0d9bf><!----><div class="items" data-v-a4b0d9bf><!--[--><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/examples" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Examples</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/previews/PR57/api" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Full API</p><!--]--></a><!----></div><!----></div><!--]--></div></section></div><!--]--><!--[--><!--]--></nav></aside><div class="VPContent has-sidebar" id="VPContent" data-v-a9a9e638 data-v-91765379><div class="VPDoc has-sidebar has-aside" data-v-91765379 data-v-83890dd9><!--[--><!--]--><div class="container" data-v-83890dd9><div class="aside" data-v-83890dd9><div class="aside-curtain" data-v-83890dd9></div><div class="aside-container" data-v-83890dd9><div class="aside-content" data-v-83890dd9><div class="VPDocAside" data-v-83890dd9 data-v-6d7b3c46><!--[--><!--]--><!--[--><!--]--><nav aria-labelledby="doc-outline-aria-label" class="VPDocAsideOutline" data-v-6d7b3c46 data-v-b38bf2ff><div class="content" data-v-b38bf2ff><div class="outline-marker" data-v-b38bf2ff></div><div aria-level="2" class="outline-title" id="doc-outline-aria-label" role="heading" data-v-b38bf2ff>On this page</div><ul class="VPDocOutlineItem root" data-v-b38bf2ff data-v-3f927ebe><!--[--><!--]--></ul></div></nav><!--[--><!--]--><div class="spacer" data-v-6d7b3c46></div><!--[--><!--]--><!----><!--[--><!--]--><!--[--><!--]--></div></div></div></div><div class="content" data-v-83890dd9><div class="content-container" data-v-83890dd9><!--[--><!--]--><main class="main" data-v-83890dd9><div style="position:relative;" class="vp-doc _MolecularEvolution_jl_previews_PR57_api" data-v-83890dd9><div><h1 id="Full-API" tabindex="-1">Full API <a class="header-anchor" href="#Full-API" aria-label="Permalink to &quot;Full API {#Full-API}&quot;">​</a></h1><h2 id="Index" tabindex="-1">Index <a class="header-anchor" href="#Index" aria-label="Permalink to &quot;Index {#Index}&quot;">​</a></h2><ul><li><a href="#MolecularEvolution.AbstractUpdate"><code>MolecularEvolution.AbstractUpdate</code></a></li><li><a href="#MolecularEvolution.AminoAcidPartition"><code>MolecularEvolution.AminoAcidPartition</code></a></li><li><a href="#MolecularEvolution.BWMModel-Union{Tuple{Vector{&lt;:M}}, Tuple{M}} where M&lt;:DiscreteStateModel"><code>MolecularEvolution.BWMModel</code></a></li><li><a href="#MolecularEvolution.BWMModel"><code>MolecularEvolution.BWMModel</code></a></li><li><a href="#MolecularEvolution.BranchlengthSampler"><code>MolecularEvolution.BranchlengthSampler</code></a></li><li><a href="#MolecularEvolution.BrownianMotion"><code>MolecularEvolution.BrownianMotion</code></a></li><li><a href="#MolecularEvolution.CATModel"><code>MolecularEvolution.CATModel</code></a></li><li><a href="#MolecularEvolution.CATPartition"><code>MolecularEvolution.CATPartition</code></a></li><li><a href="#MolecularEvolution.CladeStats"><code>MolecularEvolution.CladeStats</code></a></li><li><a href="#MolecularEvolution.CodonPartition"><code>MolecularEvolution.CodonPartition</code></a></li><li><a href="#MolecularEvolution.CovarionModel"><code>MolecularEvolution.CovarionModel</code></a></li><li><a href="#MolecularEvolution.CovarionPartition"><code>MolecularEvolution.CovarionPartition</code></a></li><li><a href="#MolecularEvolution.CustomDiscretePartition"><code>MolecularEvolution.CustomDiscretePartition</code></a></li><li><a href="#MolecularEvolution.DiagonalizedCTMC"><code>MolecularEvolution.DiagonalizedCTMC</code></a></li><li><a href="#MolecularEvolution.DiscretePartition"><code>MolecularEvolution.DiscretePartition</code></a></li><li><a href="#MolecularEvolution.GappyAminoAcidPartition"><code>MolecularEvolution.GappyAminoAcidPartition</code></a></li><li><a href="#MolecularEvolution.GappyNucleotidePartition"><code>MolecularEvolution.GappyNucleotidePartition</code></a></li><li><a href="#MolecularEvolution.GaussianPartition"><code>MolecularEvolution.GaussianPartition</code></a></li><li><a href="#MolecularEvolution.GeneralCTMC"><code>MolecularEvolution.GeneralCTMC</code></a></li><li><a href="#MolecularEvolution.InterpolatedDiscreteModel"><code>MolecularEvolution.InterpolatedDiscreteModel</code></a></li><li><a href="#MolecularEvolution.LazyDown"><code>MolecularEvolution.LazyDown</code></a></li><li><a href="#MolecularEvolution.LazyPartition"><code>MolecularEvolution.LazyPartition</code></a></li><li><a href="#MolecularEvolution.LazyUp"><code>MolecularEvolution.LazyUp</code></a></li><li><a href="#MolecularEvolution.NucleotidePartition"><code>MolecularEvolution.NucleotidePartition</code></a></li><li><a href="#MolecularEvolution.PModel"><code>MolecularEvolution.PModel</code></a></li><li><a href="#MolecularEvolution.PiQ"><code>MolecularEvolution.PiQ</code></a></li><li><a href="#MolecularEvolution.SWMModel"><code>MolecularEvolution.SWMModel</code></a></li><li><a href="#MolecularEvolution.SWMPartition"><code>MolecularEvolution.SWMPartition</code></a></li><li><a href="#MolecularEvolution.StandardUpdate"><code>MolecularEvolution.StandardUpdate</code></a></li><li><a href="#MolecularEvolution.ZeroDriftOrnsteinUhlenbeck"><code>MolecularEvolution.ZeroDriftOrnsteinUhlenbeck</code></a></li><li><a href="#Base.:==-Union{Tuple{T}, Tuple{T, T}} where T&lt;:AbstractTreeNode"><code>Base.:==</code></a></li><li><a href="#MolecularEvolution.BayesUpdate-Tuple{}"><code>MolecularEvolution.BayesUpdate</code></a></li><li><a href="#MolecularEvolution.HB98_AAfit-Tuple{Any, Any, Any}"><code>MolecularEvolution.HB98_AAfit</code></a></li><li><a href="#MolecularEvolution.HB98_F61-Tuple{Any, Any, Any}"><code>MolecularEvolution.HB98_F61</code></a></li><li><a href="#MolecularEvolution.HIPSTR"><code>MolecularEvolution.HIPSTR</code></a></li><li><a href="#MolecularEvolution.HIPSTR-Tuple{Vector{FelNode}}"><code>MolecularEvolution.HIPSTR</code></a></li><li><a href="#MolecularEvolution.MG94_F3x4-NTuple{4, Any}"><code>MolecularEvolution.MG94_F3x4</code></a></li><li><a href="#MolecularEvolution.MG94_F61-NTuple{4, Any}"><code>MolecularEvolution.MG94_F61</code></a></li><li><a href="#MolecularEvolution.MaxLikUpdate-Tuple{}"><code>MolecularEvolution.MaxLikUpdate</code></a></li><li><a href="#MolecularEvolution.SWM_prob_grid-Union{Tuple{SWMPartition{PType}}, Tuple{PType}} where PType&lt;:MultiSitePartition"><code>MolecularEvolution.SWM_prob_grid</code></a></li><li><a href="#MolecularEvolution._mapreduce-Union{Tuple{T}, Tuple{AbstractTreeNode, T, Any, Any}} where T&lt;:Function"><code>MolecularEvolution._mapreduce</code></a></li><li><a href="#MolecularEvolution.allocate!-Tuple{Any, Any}"><code>MolecularEvolution.allocate!</code></a></li><li><a href="#MolecularEvolution.backward!-Tuple{DiscretePartition, DiscretePartition, MolecularEvolution.PMatrixModel, FelNode}"><code>MolecularEvolution.backward!</code></a></li><li><a href="#MolecularEvolution.backward!"><code>MolecularEvolution.backward!</code></a></li><li><a href="#MolecularEvolution.bfs_mapreduce-Union{Tuple{T}, Tuple{AbstractTreeNode, T, Any}} where T&lt;:Function"><code>MolecularEvolution.bfs_mapreduce</code></a></li><li><a href="#MolecularEvolution.branchlength_optim!"><code>MolecularEvolution.branchlength_optim!</code></a></li><li><a href="#MolecularEvolution.branchlength_optim!-Tuple"><code>MolecularEvolution.branchlength_optim!</code></a></li><li><a href="#MolecularEvolution.branchlength_update!-Tuple{MolecularEvolution.UnivariateModifier, FelNode, Any}"><code>MolecularEvolution.branchlength_update!</code></a></li><li><a href="#MolecularEvolution.brents_method_minimize-Tuple{Any, Real, Real, Any, Real}"><code>MolecularEvolution.brents_method_minimize</code></a></li><li><a href="#MolecularEvolution.cascading_max_state_dict"><code>MolecularEvolution.cascading_max_state_dict</code></a></li><li><a href="#MolecularEvolution.cascading_max_state_dict-Tuple{FelNode, Any}"><code>MolecularEvolution.cascading_max_state_dict</code></a></li><li><a href="#MolecularEvolution.char_proportions-Tuple{Any, String}"><code>MolecularEvolution.char_proportions</code></a></li><li><a href="#MolecularEvolution.collect_clades!-Tuple{FelNode, Dict{Tuple{UInt64, UInt64}, MolecularEvolution.CladeStats}, Dict{String, Int64}}"><code>MolecularEvolution.collect_clades!</code></a></li><li><a href="#MolecularEvolution.collect_leaf_dists-Tuple{Vector{&lt;:AbstractTreeNode}}"><code>MolecularEvolution.collect_leaf_dists</code></a></li><li><a href="#MolecularEvolution.colored_seq_draw-Tuple{Any, Any, AbstractString}"><code>MolecularEvolution.colored_seq_draw</code></a></li><li><a href="#MolecularEvolution.combine!"><code>MolecularEvolution.combine!</code></a></li><li><a href="#MolecularEvolution.combine!-Tuple{DiscretePartition, DiscretePartition}"><code>MolecularEvolution.combine!</code></a></li><li><a href="#MolecularEvolution.copy_tree"><code>MolecularEvolution.copy_tree</code></a></li><li><a href="#MolecularEvolution.count_F3x4-Tuple{Array{String}}"><code>MolecularEvolution.count_F3x4</code></a></li><li><a href="#MolecularEvolution.count_F61-Tuple{Array{String}}"><code>MolecularEvolution.count_F61</code></a></li><li><a href="#MolecularEvolution.deepequals-Union{Tuple{T}, Tuple{T, T}} where T&lt;:AbstractTreeNode"><code>MolecularEvolution.deepequals</code></a></li><li><a href="#MolecularEvolution.dfs_mapreduce-Union{Tuple{T}, Tuple{AbstractTreeNode, T, Any}} where T&lt;:Function"><code>MolecularEvolution.dfs_mapreduce</code></a></li><li><a href="#MolecularEvolution.discrete_name_color_dict-Tuple{AbstractTreeNode, Any}"><code>MolecularEvolution.discrete_name_color_dict</code></a></li><li><a href="#MolecularEvolution.draw_example_tree-Tuple{}"><code>MolecularEvolution.draw_example_tree</code></a></li><li><a href="#MolecularEvolution.endpoint_conditioned_sample_state_dict-Tuple{FelNode, Any}"><code>MolecularEvolution.endpoint_conditioned_sample_state_dict</code></a></li><li><a href="#MolecularEvolution.endpoint_conditioned_sample_state_dict"><code>MolecularEvolution.endpoint_conditioned_sample_state_dict</code></a></li><li><a href="#MolecularEvolution.expected_subs_per_site-Tuple{Any, Any}"><code>MolecularEvolution.expected_subs_per_site</code></a></li><li><a href="#MolecularEvolution.felsenstein!-Tuple{FelNode, Any}"><code>MolecularEvolution.felsenstein!</code></a></li><li><a href="#MolecularEvolution.felsenstein_down!-Tuple{FelNode, Any}"><code>MolecularEvolution.felsenstein_down!</code></a></li><li><a href="#MolecularEvolution.felsenstein_roundtrip!-Tuple{FelNode, Any}"><code>MolecularEvolution.felsenstein_roundtrip!</code></a></li><li><a href="#MolecularEvolution.forward!-Tuple{DiscretePartition, DiscretePartition, MolecularEvolution.PMatrixModel, FelNode}"><code>MolecularEvolution.forward!</code></a></li><li><a href="#MolecularEvolution.forward!"><code>MolecularEvolution.forward!</code></a></li><li><a href="#MolecularEvolution.gappy_Q_from_symmetric_rate_matrix-Tuple{Any, Any, Any}"><code>MolecularEvolution.gappy_Q_from_symmetric_rate_matrix</code></a></li><li><a href="#MolecularEvolution.get_highlighter_legend-Tuple{Any}"><code>MolecularEvolution.get_highlighter_legend</code></a></li><li><a href="#MolecularEvolution.get_max_depth-Tuple{Any, Real}"><code>MolecularEvolution.get_max_depth</code></a></li><li><a href="#MolecularEvolution.get_phylo_tree-Tuple{FelNode}"><code>MolecularEvolution.get_phylo_tree</code></a></li><li><a href="#MolecularEvolution.get_phylo_tree"><code>MolecularEvolution.get_phylo_tree</code></a></li><li><a href="#MolecularEvolution.golden_section_maximize-Tuple{Any, Real, Real, Any, Real}"><code>MolecularEvolution.golden_section_maximize</code></a></li><li><a href="#MolecularEvolution.hash_clade-Tuple{BitSet}"><code>MolecularEvolution.hash_clade</code></a></li><li><a href="#MolecularEvolution.highlight_seq_draw-Tuple{Any, Any, AbstractString, Any, Any, Any}"><code>MolecularEvolution.highlight_seq_draw</code></a></li><li><a href="#MolecularEvolution.highlighter_tree_draw-NTuple{4, Any}"><code>MolecularEvolution.highlighter_tree_draw</code></a></li><li><a href="#MolecularEvolution.internal_message_init!-Tuple{FelNode, Vector{&lt;:Partition}}"><code>MolecularEvolution.internal_message_init!</code></a></li><li><a href="#MolecularEvolution.internal_message_init!-Tuple{FelNode, Partition}"><code>MolecularEvolution.internal_message_init!</code></a></li><li><a href="#MolecularEvolution.internal_nodes-Tuple{Any}"><code>MolecularEvolution.internal_nodes</code></a></li><li><a href="#MolecularEvolution.istreeconsistent-Tuple{T} where T&lt;:AbstractTreeNode"><code>MolecularEvolution.istreeconsistent</code></a></li><li><a href="#MolecularEvolution.ladder_tree_sim-Tuple{Any}"><code>MolecularEvolution.ladder_tree_sim</code></a></li><li><a href="#MolecularEvolution.lazyprep!-Tuple{FelNode, Vector{&lt;:Partition}}"><code>MolecularEvolution.lazyprep!</code></a></li><li><a href="#MolecularEvolution.lazysort!-Tuple{FelNode}"><code>MolecularEvolution.lazysort!</code></a></li><li><a href="#MolecularEvolution.leaf_distmat-Tuple{Any}"><code>MolecularEvolution.leaf_distmat</code></a></li><li><a href="#MolecularEvolution.leaf_names-Tuple{FelNode}"><code>MolecularEvolution.leaf_names</code></a></li><li><a href="#MolecularEvolution.leaf_samples-Tuple{FelNode}"><code>MolecularEvolution.leaf_samples</code></a></li><li><a href="#MolecularEvolution.leaves-Tuple{Any}"><code>MolecularEvolution.leaves</code></a></li><li><a href="#MolecularEvolution.linear_scale-NTuple{5, Any}"><code>MolecularEvolution.linear_scale</code></a></li><li><a href="#MolecularEvolution.log_likelihood-Tuple{FelNode, BranchModel}"><code>MolecularEvolution.log_likelihood</code></a></li><li><a href="#MolecularEvolution.log_likelihood!-Tuple{FelNode, Any}"><code>MolecularEvolution.log_likelihood!</code></a></li><li><a href="#MolecularEvolution.longest_path-Tuple{FelNode}"><code>MolecularEvolution.longest_path</code></a></li><li><a href="#MolecularEvolution.marginal_state_dict-Tuple{FelNode, Any}"><code>MolecularEvolution.marginal_state_dict</code></a></li><li><a href="#MolecularEvolution.marginal_state_dict"><code>MolecularEvolution.marginal_state_dict</code></a></li><li><a href="#MolecularEvolution.matrix_for_display-Tuple{Any, Any}"><code>MolecularEvolution.matrix_for_display</code></a></li><li><a href="#MolecularEvolution.metropolis_sample-Tuple{AbstractUpdate, FelNode, Any, Any}"><code>MolecularEvolution.metropolis_sample</code></a></li><li><a href="#MolecularEvolution.metropolis_sample-Tuple{FelNode, Vector{&lt;:BranchModel}, Any}"><code>MolecularEvolution.metropolis_sample</code></a></li><li><a href="#MolecularEvolution.metropolis_step-Tuple{Function, Any, Any}"><code>MolecularEvolution.metropolis_step</code></a></li><li><a href="#MolecularEvolution.midpoint-Tuple{FelNode}"><code>MolecularEvolution.midpoint</code></a></li><li><a href="#MolecularEvolution.mix-Union{Tuple{SWMPartition{PType}}, Tuple{PType}} where PType&lt;:DiscretePartition"><code>MolecularEvolution.mix</code></a></li><li><a href="#MolecularEvolution.name2node_dict-Tuple{AbstractTreeNode}"><code>MolecularEvolution.name2node_dict</code></a></li><li><a href="#MolecularEvolution.newick"><code>MolecularEvolution.newick</code></a></li><li><a href="#MolecularEvolution.newick-Tuple{AbstractTreeNode}"><code>MolecularEvolution.newick</code></a></li><li><a href="#MolecularEvolution.nni_optim!"><code>MolecularEvolution.nni_optim!</code></a></li><li><a href="#MolecularEvolution.nni_optim!-Tuple"><code>MolecularEvolution.nni_optim!</code></a></li><li><a href="#MolecularEvolution.nni_update!-Tuple{Function, FelNode, Any}"><code>MolecularEvolution.nni_update!</code></a></li><li><a href="#MolecularEvolution.node_distances-Tuple{AbstractTreeNode}"><code>MolecularEvolution.node_distances</code></a></li><li><a href="#MolecularEvolution.node_names-Tuple{FelNode}"><code>MolecularEvolution.node_names</code></a></li><li><a href="#MolecularEvolution.node_samples-Tuple{FelNode}"><code>MolecularEvolution.node_samples</code></a></li><li><a href="#MolecularEvolution.nodes-Tuple{Any}"><code>MolecularEvolution.nodes</code></a></li><li><a href="#MolecularEvolution.nonreversibleQ-Tuple{Any}"><code>MolecularEvolution.nonreversibleQ</code></a></li><li><a href="#MolecularEvolution.parent_list-Tuple{FelNode}"><code>MolecularEvolution.parent_list</code></a></li><li><a href="#MolecularEvolution.partition2obs"><code>MolecularEvolution.partition2obs</code></a></li><li><a href="#MolecularEvolution.partition2obs-Tuple{DiscretePartition, String}"><code>MolecularEvolution.partition2obs</code></a></li><li><a href="#MolecularEvolution.plot_multiple_trees"><code>MolecularEvolution.plot_multiple_trees</code></a></li><li><a href="#MolecularEvolution.plot_multiple_trees-Tuple{Any, Any}"><code>MolecularEvolution.plot_multiple_trees</code></a></li><li><a href="#MolecularEvolution.populate_tree!-Tuple{FelNode, Partition, Any, Any}"><code>MolecularEvolution.populate_tree!</code></a></li><li><a href="#MolecularEvolution.populate_tree!"><code>MolecularEvolution.populate_tree!</code></a></li><li><a href="#MolecularEvolution.promote_internal-Tuple{FelNode}"><code>MolecularEvolution.promote_internal</code></a></li><li><a href="#MolecularEvolution.quadratic_CI-Tuple{Function, Vector, Int64}"><code>MolecularEvolution.quadratic_CI</code></a></li><li><a href="#MolecularEvolution.quadratic_CI-Tuple{Vector, Vector}"><code>MolecularEvolution.quadratic_CI</code></a></li><li><a href="#MolecularEvolution.read_fasta-Tuple{String}"><code>MolecularEvolution.read_fasta</code></a></li><li><a href="#MolecularEvolution.read_fasta"><code>MolecularEvolution.read_fasta</code></a></li><li><a href="#MolecularEvolution.read_newick_tree-Tuple{String}"><code>MolecularEvolution.read_newick_tree</code></a></li><li><a href="#MolecularEvolution.read_newick_tree"><code>MolecularEvolution.read_newick_tree</code></a></li><li><a href="#MolecularEvolution.refresh!-Tuple{FelNode, Any}"><code>MolecularEvolution.refresh!</code></a></li><li><a href="#MolecularEvolution.reversibleQ"><code>MolecularEvolution.reversibleQ</code></a></li><li><a href="#MolecularEvolution.reversibleQ-Tuple{Any, Any}"><code>MolecularEvolution.reversibleQ</code></a></li><li><a href="#MolecularEvolution.root2tip_distances-Tuple{AbstractTreeNode}"><code>MolecularEvolution.root2tip_distances</code></a></li><li><a href="#MolecularEvolution.root_optim!"><code>MolecularEvolution.root_optim!</code></a></li><li><a href="#MolecularEvolution.root_optim!-Tuple{FelNode, Any}"><code>MolecularEvolution.root_optim!</code></a></li><li><a href="#MolecularEvolution.root_update!-Tuple{RootUpdate, FelNode, Any}"><code>MolecularEvolution.root_update!</code></a></li><li><a href="#MolecularEvolution.sample_down!-Tuple{FelNode, Any, Any}"><code>MolecularEvolution.sample_down!</code></a></li><li><a href="#MolecularEvolution.sample_down!"><code>MolecularEvolution.sample_down!</code></a></li><li><a href="#MolecularEvolution.sample_from_message!-Tuple{Vector{&lt;:Partition}}"><code>MolecularEvolution.sample_from_message!</code></a></li><li><a href="#MolecularEvolution.savefig_tweakSVG-Tuple{Any, Plots.Plot}"><code>MolecularEvolution.savefig_tweakSVG</code></a></li><li><a href="#MolecularEvolution.savefig_tweakSVG"><code>MolecularEvolution.savefig_tweakSVG</code></a></li><li><a href="#MolecularEvolution.savefig_tweakSVG-Tuple{Any, Context}"><code>MolecularEvolution.savefig_tweakSVG</code></a></li><li><a href="#MolecularEvolution.shortest_path_between_nodes-Tuple{FelNode, FelNode}"><code>MolecularEvolution.shortest_path_between_nodes</code></a></li><li><a href="#MolecularEvolution.sibling_inds-Tuple{AbstractTreeNode}"><code>MolecularEvolution.sibling_inds</code></a></li><li><a href="#MolecularEvolution.siblings-Tuple{AbstractTreeNode}"><code>MolecularEvolution.siblings</code></a></li><li><a href="#MolecularEvolution.sim_tree-Tuple{Int64, Any, Any}"><code>MolecularEvolution.sim_tree</code></a></li><li><a href="#MolecularEvolution.sim_tree-Tuple{}"><code>MolecularEvolution.sim_tree</code></a></li><li><a href="#MolecularEvolution.sim_tree"><code>MolecularEvolution.sim_tree</code></a></li><li><a href="#MolecularEvolution.simple_radial_tree_plot-Tuple{FelNode}"><code>MolecularEvolution.simple_radial_tree_plot</code></a></li><li><a href="#MolecularEvolution.simple_tree_draw-Tuple{FelNode}"><code>MolecularEvolution.simple_tree_draw</code></a></li><li><a href="#MolecularEvolution.standard_tree_sim-Tuple{Any}"><code>MolecularEvolution.standard_tree_sim</code></a></li><li><a href="#MolecularEvolution.total_LL-Tuple{Partition}"><code>MolecularEvolution.total_LL</code></a></li><li><a href="#MolecularEvolution.tree2distances-Tuple{AbstractTreeNode}"><code>MolecularEvolution.tree2distances</code></a></li><li><a href="#MolecularEvolution.tree2shared_branch_lengths-Tuple{AbstractTreeNode}"><code>MolecularEvolution.tree2shared_branch_lengths</code></a></li><li><a href="#MolecularEvolution.tree_draw-Tuple{FelNode}"><code>MolecularEvolution.tree_draw</code></a></li><li><a href="#MolecularEvolution.tree_draw"><code>MolecularEvolution.tree_draw</code></a></li><li><a href="#MolecularEvolution.tree_polish!"><code>MolecularEvolution.tree_polish!</code></a></li><li><a href="#MolecularEvolution.tree_polish!-Tuple{Any, Any}"><code>MolecularEvolution.tree_polish!</code></a></li><li><a href="#MolecularEvolution.unc2probvec"><code>MolecularEvolution.unc2probvec</code></a></li><li><a href="#MolecularEvolution.unc2probvec-Tuple{Any}"><code>MolecularEvolution.unc2probvec</code></a></li><li><a href="#MolecularEvolution.univariate_maximize-Tuple{Any, Real, Real, Any, GoldenSectionOpt, Real}"><code>MolecularEvolution.univariate_maximize</code></a></li><li><a href="#MolecularEvolution.univariate_maximize-Tuple{Any, Real, Real, Any, BrentsMethodOpt, Real}"><code>MolecularEvolution.univariate_maximize</code></a></li><li><a href="#MolecularEvolution.univariate_sampler-Tuple{Any, BranchlengthSampler, Any}"><code>MolecularEvolution.univariate_sampler</code></a></li><li><a href="#MolecularEvolution.values_from_phylo_tree-Tuple{Any, Any}"><code>MolecularEvolution.values_from_phylo_tree</code></a></li><li><a href="#MolecularEvolution.values_from_phylo_tree"><code>MolecularEvolution.values_from_phylo_tree</code></a></li><li><a href="#MolecularEvolution.weightEM-Tuple{Matrix{Float64}, Any}"><code>MolecularEvolution.weightEM</code></a></li><li><a href="#MolecularEvolution.write_fasta"><code>MolecularEvolution.write_fasta</code></a></li><li><a href="#MolecularEvolution.write_fasta-Tuple{String, Vector{String}}"><code>MolecularEvolution.write_fasta</code></a></li><li><a href="#MolecularEvolution.write_nexus"><code>MolecularEvolution.write_nexus</code></a></li><li><a href="#MolecularEvolution.write_nexus-Tuple{String, FelNode}"><code>MolecularEvolution.write_nexus</code></a></li></ul><h2 id="Docstrings" tabindex="-1">Docstrings <a class="header-anchor" href="#Docstrings" aria-label="Permalink to &quot;Docstrings {#Docstrings}&quot;">​</a></h2><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.AbstractUpdate" href="#MolecularEvolution.AbstractUpdate"><span class="jlbinding">MolecularEvolution.AbstractUpdate</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Summary</strong></p><p><code>abstract type AbstractUpdate &lt;: Function</code></p><p>A callable type that typically takes <code>(tree::FelNode, models; partition_list=1:length(tree.message))</code>, updates <code>tree</code> and <code>models</code>, and returns the updated <code>tree</code> and <code>models</code>.</p><p><strong>Example</strong></p><p>Define a new subtype, where <code>foo</code> and <code>bar</code> are arbitrary updating functions</p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">struct</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> MyUpdate </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> AbstractUpdate</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> end</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> (update</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">MyUpdate</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; partition_list</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message))</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    tree, models </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> foo</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree, models, partition_list</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">partition_list)</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    tree, models </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> BayesUpdate</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(nni</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)(tree, models, partition_list</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">partition_list)</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    tree, models </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> bar</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree, models, partition_list</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">partition_list)</span></span>
<span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">    return</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> tree, models</span></span>
<span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">end</span></span></code></pre></div><p>See also: <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.StandardUpdate"><code>StandardUpdate</code></a></p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/AbstractUpdate.jl#L1-L20" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.AminoAcidPartition" href="#MolecularEvolution.AminoAcidPartition"><span class="jlbinding">MolecularEvolution.AminoAcidPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">mutable struct</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> AminoAcidPartition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscretePartition</span></span></code></pre></div><p><strong>Constructors</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>AminoAcidPartition(sites)</span></span>
<span class="line"><span>AminoAcidPartition(freq_vec::Vector{Float64}, sites::Int64)</span></span>
<span class="line"><span>AminoAcidPartition(state, states, sites, scaling)</span></span></code></pre></div><p><strong>Description</strong></p><p>A <code>DiscretePartition</code> for amino acid sequences with 20 states representing the standard amino acids.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/discrete_partitions.jl#L125-L134" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.BWMModel" href="#MolecularEvolution.BWMModel"><span class="jlbinding">MolecularEvolution.BWMModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">mutable struct</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> BWMModel{M} </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscreteStateModel</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> M </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscreteStateModel</span></span></code></pre></div><p><strong>Fields</strong></p><ul><li><p><code>models::Vector{&lt;:M}</code>: A vector of models.</p></li><li><p><code>weights::Vector{Float64}</code>: A vector of weights.</p></li></ul><p><strong>Description</strong></p><p>Branch-wise mixture model.</p><div class="tip custom-block"><p class="custom-block-title">Note</p><p><code>forward!</code> and <code>backward!</code> are currently only defined for <code>M&lt;:PMatrixModel</code>.</p></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/compound_models/bwm.jl#L2-L13" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.BWMModel-Union{Tuple{Vector{&lt;:M}}, Tuple{M}} where M&lt;:DiscreteStateModel" href="#MolecularEvolution.BWMModel-Union{Tuple{Vector{&lt;:M}}, Tuple{M}} where M&lt;:DiscreteStateModel"><span class="jlbinding">MolecularEvolution.BWMModel</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">BWMModel{M}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:M}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> M </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscreteStateModel</span></span></code></pre></div><p>Convenience constructor where the weights are uniform.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/compound_models/bwm.jl#L19-L23" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.BranchlengthSampler" href="#MolecularEvolution.BranchlengthSampler"><span class="jlbinding">MolecularEvolution.BranchlengthSampler</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">BranchlengthSampler</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">A type that allows you to specify a additive proposal </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> in</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> the log domain and a prior distrubution over the log of the branchlengths. It also stores </span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;">`acc_ratio`</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> which is a tuple of </span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;">`(ratio, total, #acceptances)`</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;"> `ratio::Float64`</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> is the acceptance ratio, </span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;">`total::Int64`</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> is the total number of proposals, and </span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;">`#acceptances::Int64`</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> is the number of acceptances.</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/simple_sample.jl#L6-L10" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.BrownianMotion" href="#MolecularEvolution.BrownianMotion"><span class="jlbinding">MolecularEvolution.BrownianMotion</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">BrownianMotion</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(mean_drift</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, var_drift</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>A 1D continuous Brownian motion model with mean drift and variance drift.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/continuous_models/brownian_motion.jl#L2-L6" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CATModel" href="#MolecularEvolution.CATModel"><span class="jlbinding">MolecularEvolution.CATModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CATModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:BranchModel}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>CAT is something where you split the sites up, and assign each site to a different model (whose &quot;data&quot; gets stored in a contiguous block of memory).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/compound_models/cat.jl#L2-L6" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CATPartition" href="#MolecularEvolution.CATPartition"><span class="jlbinding">MolecularEvolution.CATPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CATPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(part_inds</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Vector{Int}}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CATPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(part_inds</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Vector{Int}}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, parts</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{PType}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p><strong>Description</strong></p><p>A partition for the <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.CATModel"><code>CATModel</code></a>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/compound_models/cat.jl#L15-L23" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CladeStats" href="#MolecularEvolution.CladeStats"><span class="jlbinding">MolecularEvolution.CladeStats</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p>Store statistics about a clade: its frequency and observed child pairs.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/HIPSTR.jl#L164-L166" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CodonPartition" href="#MolecularEvolution.CodonPartition"><span class="jlbinding">MolecularEvolution.CodonPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">mutable struct</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> CodonPartition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscretePartition</span></span></code></pre></div><p><strong>Constructors</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>CodonPartition(sites; code = universal_code)</span></span>
<span class="line"><span>CodonPartition(state, states, sites, scaling; code = universal_code)</span></span></code></pre></div><p><strong>Description</strong></p><p>A <code>DiscretePartition</code> where every state represents a codon. Can be customized to use different genetic codes.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/codon_models.jl#L349-L357" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CovarionModel" href="#MolecularEvolution.CovarionModel"><span class="jlbinding">MolecularEvolution.CovarionModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CovarionModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:DiscreteStateModel}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, inter_transition_rates</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Matrix{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CovarionModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:DiscreteStateModel}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, inter_transition_rate</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p><strong>Description</strong></p><p>The covarion model.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/compound_models/covarion.jl#L2-L10" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CovarionPartition" href="#MolecularEvolution.CovarionPartition"><span class="jlbinding">MolecularEvolution.CovarionPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CovarionPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(states,sites,models,t)</span></span></code></pre></div><p>A partition for the <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.CovarionModel"><code>CovarionModel</code></a>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/compound_models/covarion.jl#L47-L51" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CustomDiscretePartition" href="#MolecularEvolution.CustomDiscretePartition"><span class="jlbinding">MolecularEvolution.CustomDiscretePartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">mutable struct</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> CustomDiscretePartition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscretePartition</span></span></code></pre></div><p><strong>Constructors</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>CustomDiscretePartition(states, sites)</span></span>
<span class="line"><span>CustomDiscretePartition(freq_vec::Vector{Float64}, sites::Int64)</span></span>
<span class="line"><span>CustomDiscretePartition(state, states, sites, scaling)</span></span></code></pre></div><p><strong>Description</strong></p><p>A general-purpose <code>DiscretePartition</code> that can handle any number of states. Useful for custom discrete state spaces that don&#39;t fit into the predefined partition types.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/discrete_partitions.jl#L35-L45" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.DiagonalizedCTMC" href="#MolecularEvolution.DiagonalizedCTMC"><span class="jlbinding">MolecularEvolution.DiagonalizedCTMC</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">DiagonalizedCTMC</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(Q</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Array{Float64,2}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">DiagonalizedCTMC</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(Qpre</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Array{Float64,2}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">pi</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p><strong>Description</strong></p><p>Takes in a Q matrix (which can be multiplied onto row-wise by <code>pi</code>) and diagonalizes it. When computing <mjx-container class="MathJax" jax="SVG" style="direction:ltr;position:relative;"><svg style="overflow:visible;min-height:1px;min-width:1px;vertical-align:-0.025ex;" xmlns="http://www.w3.org/2000/svg" width="3.085ex" height="1.972ex" role="img" focusable="false" viewBox="0 -860.8 1363.6 871.8" aria-hidden="true"><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msup"><g data-mml-node="mi"><path data-c="1D452" d="M39 168Q39 225 58 272T107 350T174 402T244 433T307 442H310Q355 442 388 420T421 355Q421 265 310 237Q261 224 176 223Q139 223 138 221Q138 219 132 186T125 128Q125 81 146 54T209 26T302 45T394 111Q403 121 406 121Q410 121 419 112T429 98T420 82T390 55T344 24T281 -1T205 -11Q126 -11 83 42T39 168ZM373 353Q367 405 305 405Q272 405 244 391T199 357T170 316T154 280T149 261Q149 260 169 260Q282 260 327 284T373 353Z" style="stroke-width:3;"></path></g><g data-mml-node="TeXAtom" transform="translate(499,363) scale(0.707)" data-mjx-texclass="ORD"><g data-mml-node="mi"><path data-c="1D444" d="M399 -80Q399 -47 400 -30T402 -11V-7L387 -11Q341 -22 303 -22Q208 -22 138 35T51 201Q50 209 50 244Q50 346 98 438T227 601Q351 704 476 704Q514 704 524 703Q621 689 680 617T740 435Q740 255 592 107Q529 47 461 16L444 8V3Q444 2 449 -24T470 -66T516 -82Q551 -82 583 -60T625 -3Q631 11 638 11Q647 11 649 2Q649 -6 639 -34T611 -100T557 -165T481 -194Q399 -194 399 -87V-80ZM636 468Q636 523 621 564T580 625T530 655T477 665Q429 665 379 640Q277 591 215 464T153 216Q153 110 207 59Q231 38 236 38V46Q236 86 269 120T347 155Q372 155 390 144T417 114T429 82T435 55L448 64Q512 108 557 185T619 334T636 468ZM314 18Q362 18 404 39L403 49Q399 104 366 115Q354 117 347 117Q344 117 341 117T337 118Q317 118 296 98T274 52Q274 18 314 18Z" style="stroke-width:3;"></path></g><g data-mml-node="mi" transform="translate(791,0)"><path data-c="1D461" d="M26 385Q19 392 19 395Q19 399 22 411T27 425Q29 430 36 430T87 431H140L159 511Q162 522 166 540T173 566T179 586T187 603T197 615T211 624T229 626Q247 625 254 615T261 596Q261 589 252 549T232 470L222 433Q222 431 272 431H323Q330 424 330 420Q330 398 317 385H210L174 240Q135 80 135 68Q135 26 162 26Q197 26 230 60T283 144Q285 150 288 151T303 153H307Q322 153 322 145Q322 142 319 133Q314 117 301 95T267 48T216 6T155 -11Q125 -11 98 4T59 56Q57 64 57 83V101L92 241Q127 382 128 383Q128 385 77 385H26Z" style="stroke-width:3;"></path></g></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline" style="top:0px;left:0px;clip:rect(1px, 1px, 1px, 1px);-webkit-touch-callout:none;-webkit-user-select:none;-khtml-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;position:absolute;padding:1px 0px 0px 0px;border:0px;display:block;width:auto;overflow:hidden;"><math xmlns="http://www.w3.org/1998/Math/MathML"><msup><mi>e</mi><mrow data-mjx-texclass="ORD"><mi>Q</mi><mi>t</mi></mrow></msup></math></mjx-assistive-mml></mjx-container> (for different <mjx-container class="MathJax" jax="SVG" style="direction:ltr;position:relative;"><svg style="overflow:visible;min-height:1px;min-width:1px;vertical-align:-0.025ex;" xmlns="http://www.w3.org/2000/svg" width="0.817ex" height="1.441ex" role="img" focusable="false" viewBox="0 -626 361 637" aria-hidden="true"><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><path data-c="1D461" d="M26 385Q19 392 19 395Q19 399 22 411T27 425Q29 430 36 430T87 431H140L159 511Q162 522 166 540T173 566T179 586T187 603T197 615T211 624T229 626Q247 625 254 615T261 596Q261 589 252 549T232 470L222 433Q222 431 272 431H323Q330 424 330 420Q330 398 317 385H210L174 240Q135 80 135 68Q135 26 162 26Q197 26 230 60T283 144Q285 150 288 151T303 153H307Q322 153 322 145Q322 142 319 133Q314 117 301 95T267 48T216 6T155 -11Q125 -11 98 4T59 56Q57 64 57 83V101L92 241Q127 382 128 383Q128 385 77 385H26Z" style="stroke-width:3;"></path></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline" style="top:0px;left:0px;clip:rect(1px, 1px, 1px, 1px);-webkit-touch-callout:none;-webkit-user-select:none;-khtml-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;position:absolute;padding:1px 0px 0px 0px;border:0px;display:block;width:auto;overflow:hidden;"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>t</mi></math></mjx-assistive-mml></mjx-container>s), we now only need to exponentiate the eigenvalues, and perform the two change-of-basis matrix multiplications.</p><div class="warning custom-block"><p class="custom-block-title">Warning</p><p>Construction fails if <code>Q</code> has complex eigenvalues.</p></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/pmatrix_models/DiagonalizedCTMC.jl#L5-L17" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.DiscretePartition" href="#MolecularEvolution.DiscretePartition"><span class="jlbinding">MolecularEvolution.DiscretePartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">abstract type</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> DiscretePartition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> MultiSitePartition</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> end</span></span></code></pre></div><p>Represents a <code>states</code>-by-<code>sites</code> matrix of probabilities. The following fields are loosely required:</p><ul><li><p><code>state</code>: A matrix of probabilities that are site-wise normalized.</p></li><li><p><code>states</code>: The number of states.</p></li><li><p><code>sites</code>: The number of sites.</p></li><li><p><code>scaling</code>: A vector of log-domain probability scaling factors, one per site.</p></li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/discrete_partitions.jl#L5-L13" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.GappyAminoAcidPartition" href="#MolecularEvolution.GappyAminoAcidPartition"><span class="jlbinding">MolecularEvolution.GappyAminoAcidPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">mutable struct</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> GappyAminoAcidPartition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscretePartition</span></span></code></pre></div><p><strong>Constructors</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>GappyAminoAcidPartition(sites)</span></span>
<span class="line"><span>GappyAminoAcidPartition(freq_vec::Vector{Float64}, sites::Int64)</span></span>
<span class="line"><span>GappyAminoAcidPartition(state, states, sites, scaling)</span></span></code></pre></div><p><strong>Description</strong></p><p>A <code>DiscretePartition</code> for amino acid sequences with 21 states (20 standard amino acids plus &#39;-&#39; for gaps).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/discrete_partitions.jl#L155-L164" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.GappyNucleotidePartition" href="#MolecularEvolution.GappyNucleotidePartition"><span class="jlbinding">MolecularEvolution.GappyNucleotidePartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">mutable struct</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> GappyNucleotidePartition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscretePartition</span></span></code></pre></div><p><strong>Constructors</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>GappyNucleotidePartition(sites)</span></span>
<span class="line"><span>GappyNucleotidePartition(freq_vec::Vector{Float64}, sites::Int64)</span></span>
<span class="line"><span>GappyNucleotidePartition(state, states, sites, scaling)</span></span></code></pre></div><p><strong>Description</strong></p><p>A <code>DiscretePartition</code> for nucleotide sequences with 5 states (A, C, G, T, -), where &#39;-&#39; represents a gap.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/discrete_partitions.jl#L95-L104" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.GaussianPartition" href="#MolecularEvolution.GaussianPartition"><span class="jlbinding">MolecularEvolution.GaussianPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">GaussianPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(mean</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, var</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, norm_const</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">GaussianPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(mean</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, var</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#6A737D;--shiki-dark:#6A737D;"># norm_const defaults to 0.0</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">GaussianPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">() </span><span style="--shiki-light:#6A737D;--shiki-dark:#6A737D;"># mean, var, norm_const default to 0.0, 1.0, 0.0 respectively</span></span></code></pre></div><p><strong>Description</strong></p><p>A partition representing a (not necessarily normalized) Gaussian distribution. <code>norm_const</code> is the log-domain normalization constant.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/continuous_models/gaussian_partition.jl#L3-L12" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.GeneralCTMC" href="#MolecularEvolution.GeneralCTMC"><span class="jlbinding">MolecularEvolution.GeneralCTMC</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> GeneralCTMC</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(Q</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Array{Float64,2}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Wraps a Q matrix and will compute <mjx-container class="MathJax" jax="SVG" style="direction:ltr;position:relative;"><svg style="overflow:visible;min-height:1px;min-width:1px;vertical-align:-0.025ex;" xmlns="http://www.w3.org/2000/svg" width="3.085ex" height="1.972ex" role="img" focusable="false" viewBox="0 -860.8 1363.6 871.8" aria-hidden="true"><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="msup"><g data-mml-node="mi"><path data-c="1D452" d="M39 168Q39 225 58 272T107 350T174 402T244 433T307 442H310Q355 442 388 420T421 355Q421 265 310 237Q261 224 176 223Q139 223 138 221Q138 219 132 186T125 128Q125 81 146 54T209 26T302 45T394 111Q403 121 406 121Q410 121 419 112T429 98T420 82T390 55T344 24T281 -1T205 -11Q126 -11 83 42T39 168ZM373 353Q367 405 305 405Q272 405 244 391T199 357T170 316T154 280T149 261Q149 260 169 260Q282 260 327 284T373 353Z" style="stroke-width:3;"></path></g><g data-mml-node="TeXAtom" transform="translate(499,363) scale(0.707)" data-mjx-texclass="ORD"><g data-mml-node="mi"><path data-c="1D444" d="M399 -80Q399 -47 400 -30T402 -11V-7L387 -11Q341 -22 303 -22Q208 -22 138 35T51 201Q50 209 50 244Q50 346 98 438T227 601Q351 704 476 704Q514 704 524 703Q621 689 680 617T740 435Q740 255 592 107Q529 47 461 16L444 8V3Q444 2 449 -24T470 -66T516 -82Q551 -82 583 -60T625 -3Q631 11 638 11Q647 11 649 2Q649 -6 639 -34T611 -100T557 -165T481 -194Q399 -194 399 -87V-80ZM636 468Q636 523 621 564T580 625T530 655T477 665Q429 665 379 640Q277 591 215 464T153 216Q153 110 207 59Q231 38 236 38V46Q236 86 269 120T347 155Q372 155 390 144T417 114T429 82T435 55L448 64Q512 108 557 185T619 334T636 468ZM314 18Q362 18 404 39L403 49Q399 104 366 115Q354 117 347 117Q344 117 341 117T337 118Q317 118 296 98T274 52Q274 18 314 18Z" style="stroke-width:3;"></path></g><g data-mml-node="mi" transform="translate(791,0)"><path data-c="1D461" d="M26 385Q19 392 19 395Q19 399 22 411T27 425Q29 430 36 430T87 431H140L159 511Q162 522 166 540T173 566T179 586T187 603T197 615T211 624T229 626Q247 625 254 615T261 596Q261 589 252 549T232 470L222 433Q222 431 272 431H323Q330 424 330 420Q330 398 317 385H210L174 240Q135 80 135 68Q135 26 162 26Q197 26 230 60T283 144Q285 150 288 151T303 153H307Q322 153 322 145Q322 142 319 133Q314 117 301 95T267 48T216 6T155 -11Q125 -11 98 4T59 56Q57 64 57 83V101L92 241Q127 382 128 383Q128 385 77 385H26Z" style="stroke-width:3;"></path></g></g></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline" style="top:0px;left:0px;clip:rect(1px, 1px, 1px, 1px);-webkit-touch-callout:none;-webkit-user-select:none;-khtml-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;position:absolute;padding:1px 0px 0px 0px;border:0px;display:block;width:auto;overflow:hidden;"><math xmlns="http://www.w3.org/1998/Math/MathML"><msup><mi>e</mi><mrow data-mjx-texclass="ORD"><mi>Q</mi><mi>t</mi></mrow></msup></math></mjx-assistive-mml></mjx-container> (for different <mjx-container class="MathJax" jax="SVG" style="direction:ltr;position:relative;"><svg style="overflow:visible;min-height:1px;min-width:1px;vertical-align:-0.025ex;" xmlns="http://www.w3.org/2000/svg" width="0.817ex" height="1.441ex" role="img" focusable="false" viewBox="0 -626 361 637" aria-hidden="true"><g stroke="currentColor" fill="currentColor" stroke-width="0" transform="scale(1,-1)"><g data-mml-node="math"><g data-mml-node="mi"><path data-c="1D461" d="M26 385Q19 392 19 395Q19 399 22 411T27 425Q29 430 36 430T87 431H140L159 511Q162 522 166 540T173 566T179 586T187 603T197 615T211 624T229 626Q247 625 254 615T261 596Q261 589 252 549T232 470L222 433Q222 431 272 431H323Q330 424 330 420Q330 398 317 385H210L174 240Q135 80 135 68Q135 26 162 26Q197 26 230 60T283 144Q285 150 288 151T303 153H307Q322 153 322 145Q322 142 319 133Q314 117 301 95T267 48T216 6T155 -11Q125 -11 98 4T59 56Q57 64 57 83V101L92 241Q127 382 128 383Q128 385 77 385H26Z" style="stroke-width:3;"></path></g></g></g></svg><mjx-assistive-mml unselectable="on" display="inline" style="top:0px;left:0px;clip:rect(1px, 1px, 1px, 1px);-webkit-touch-callout:none;-webkit-user-select:none;-khtml-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;position:absolute;padding:1px 0px 0px 0px;border:0px;display:block;width:auto;overflow:hidden;"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>t</mi></math></mjx-assistive-mml></mjx-container>s) naively by <code>exp</code>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/pmatrix_models/GeneralCTMC.jl#L1-L5" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.InterpolatedDiscreteModel" href="#MolecularEvolution.InterpolatedDiscreteModel"><span class="jlbinding">MolecularEvolution.InterpolatedDiscreteModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">InterpolatedDiscreteModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(siz</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, generator, tvec</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">InterpolatedDiscreteModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(Pvec</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Array{Float64,3}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, tvec</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p><code>generator</code> is a function that takes a time value <code>t</code> and returns a P matrix.</p><p><strong>Description</strong></p><p>Stores a number (<code>siz</code>) of P matrices, and the time values to which they correspond. For a requested t, the returned P matrix is (element-wise linearly) interpolated between it&#39;s two neighbours.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/interpolated_discrete_model.jl#L5-L15" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.LazyDown" href="#MolecularEvolution.LazyDown"><span class="jlbinding">MolecularEvolution.LazyDown</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>LazyDown(stores_obs)</span></span>
<span class="line"><span>LazyDown() = LazyDown(x::FelNode -&gt; true)</span></span></code></pre></div><p><strong>Description</strong></p><p>Indicate that we want to do a downward pass, e.g. <code>sample_down!</code>. The function passed to the constructor takes a <code>node::FelNode</code> as input and returns a <code>Bool</code> that decides if <code>node</code> stores its observations.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/lazy_models/lazy_partition.jl#L51-L59" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.LazyPartition" href="#MolecularEvolution.LazyPartition"><span class="jlbinding">MolecularEvolution.LazyPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructor</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>LazyPartition{PType}()</span></span></code></pre></div><p>Initialize an empty <code>LazyPartition</code> that is meant for wrapping a partition of type <code>PType</code>.</p><p><strong>Description</strong></p><p>With this data structure, you can wrap a partition of choice. The idea is that in some message passing algorithms, there is only a wave of partitions which need to actualize. For instance, a wave following a root-leaf path, or a depth-first traversal. In which case, we can be more economical with our memory consumption. With a worst case memory complexity of O(log(n)), where n is the number of nodes, functionality is provided for:</p><ul><li><p><code>log_likelihood!</code></p></li><li><p><code>felsenstein!</code></p></li><li><p><code>sample_down!</code></p></li></ul><div class="tip custom-block"><p class="custom-block-title">Note</p><p>For successive <code>felsenstein!</code> calls, we need to extract the information at the root somehow after each call. This can be done with e.g. <code>total_LL</code> or <code>site_LLs</code>.</p></div><p><strong>Further requirements</strong></p><p>Suppose you want to wrap a partition of <code>PType</code> with <code>LazyPartition</code>:</p><ul><li><p>If you&#39;re calling <code>log_likelihood!</code> and <code>felsenstein!</code>:</p><ul><li><code>obs2partition!(partition::PType, obs)</code> that transforms an observation to a partition.</li></ul></li><li><p>If you&#39;re calling <code>sample_down!</code>:</p><ul><li><code>partition2obs(partition::PType)</code> that returns the most likely state from a partition, inverts <code>obs2partition!</code>.</li></ul></li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/lazy_models/lazy_partition.jl#L2-L25" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.LazyUp" href="#MolecularEvolution.LazyUp"><span class="jlbinding">MolecularEvolution.LazyUp</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructor</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>LazyUp()</span></span></code></pre></div><p><strong>Description</strong></p><p>Indicate that we want to do an upward pass, e.g. <code>felsenstein!</code>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/lazy_models/lazy_partition.jl#L42-L48" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.NucleotidePartition" href="#MolecularEvolution.NucleotidePartition"><span class="jlbinding">MolecularEvolution.NucleotidePartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">mutable struct</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> NucleotidePartition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscretePartition</span></span></code></pre></div><p><strong>Constructors</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>NucleotidePartition(sites)</span></span>
<span class="line"><span>NucleotidePartition(freq_vec::Vector{Float64}, sites::Int64)</span></span>
<span class="line"><span>NucleotidePartition(state, states, sites, scaling)</span></span></code></pre></div><p><strong>Description</strong></p><p>A <code>DiscretePartition</code> specifically for nucleotide sequences with 4 states (A, C, G, T).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/discrete_partitions.jl#L65-L74" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.PModel" href="#MolecularEvolution.PModel"><span class="jlbinding">MolecularEvolution.PModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">PModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(P</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Array{Float64,2}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>A P matrix.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/pmatrix_models/PModel.jl#L2-L6" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.PiQ" href="#MolecularEvolution.PiQ"><span class="jlbinding">MolecularEvolution.PiQ</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">PiQ</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(r</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">pi</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; normalize</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">PiQ</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">pi</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; normalize</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p><strong>Description</strong></p><p>The F81 substitution model, but for general dimensions. <a href="https://www.diva-portal.org/smash/get/diva2:1878793/FULLTEXT01.pdf" target="_blank" rel="noreferrer">https://www.diva-portal.org/smash/get/diva2:1878793/FULLTEXT01.pdf</a></p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/PiQ.jl#L4-L13" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.SWMModel" href="#MolecularEvolution.SWMModel"><span class="jlbinding">MolecularEvolution.SWMModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:BranchModel}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(model</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">M</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, rs</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> {M </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> BranchModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">}</span></span></code></pre></div><p><strong>Description</strong></p><p>A site-wise mixture model, for site-to-site &quot;random effects&quot; rate variation.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/compound_models/swm.jl#L17-L25" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.SWMPartition" href="#MolecularEvolution.SWMPartition"><span class="jlbinding">MolecularEvolution.SWMPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(parts</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{PType}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> {PType </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> MultiSitePartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">}</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(part</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">PType</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, n_parts</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> {PType </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> MultiSitePartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">}</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(parts</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{PType}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, weights</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, sites</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, states</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> {PType </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> MultiSitePartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">}</span></span></code></pre></div><p><strong>Description</strong></p><p>A site-wise mixture partition for the <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.SWMModel"><code>SWMModel</code></a>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/compound_models/swm.jl#L46-L55" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.StandardUpdate" href="#MolecularEvolution.StandardUpdate"><span class="jlbinding">MolecularEvolution.StandardUpdate</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Summary</strong></p><p><code>struct StandardUpdate &lt;: AbstractUpdate</code></p><p>A standard update can be a family of calls to <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.nni_update!-Tuple{Function, FelNode, Any}"><code>nni_update!</code></a>, <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.branchlength_update!-Tuple{MolecularEvolution.UnivariateModifier, FelNode, Any}"><code>branchlength_update!</code></a>, <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.root_update!-Tuple{RootUpdate, FelNode, Any}"><code>root_update!</code></a>, and model updates.</p><p><strong>Constructor</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>StandardUpdate(</span></span>
<span class="line"><span>    nni::Int,</span></span>
<span class="line"><span>    branchlength::Int,</span></span>
<span class="line"><span>    root::Int,</span></span>
<span class="line"><span>    models::Int,</span></span>
<span class="line"><span>    refresh::Bool,</span></span>
<span class="line"><span>    nni_selection::Function,</span></span>
<span class="line"><span>    branchlength_modifier::UnivariateModifier,</span></span>
<span class="line"><span>    root_update::RootUpdate,</span></span>
<span class="line"><span>    models_update::ModelsUpdate</span></span>
<span class="line"><span>)</span></span></code></pre></div><p><strong>Arguments</strong></p><ul><li><p><code>nni::Int</code>: the number of times to update the tree by <code>nni_update!</code></p></li><li><p><code>branchlength::Int</code>: the number of times to update the tree by <code>branchlength_update!</code></p></li><li><p><code>root::Int</code>: the number of times to update the tree by <code>root_update!</code></p></li><li><p><code>models::Int</code>: the number of times to update the model</p></li><li><p><code>refresh::Bool</code>: whether to refresh the messages in tree between update operations to ensure message consistency</p></li><li><p><code>nni_selection::Function</code>: the function that selects between nni configurations</p></li><li><p><code>branchlength_modifier::UnivariateModifier</code>: the modifier to update a branchlength by <code>branchlength_update!</code></p></li><li><p><code>root_update::RootUpdate</code>: updates the root by <code>root_update!</code></p></li><li><p><code>models_update::ModelsUpdate</code>: updates the model parameters</p></li></ul><p>See also: <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.BayesUpdate-Tuple{}"><code>BayesUpdate</code></a>, <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.MaxLikUpdate-Tuple{}"><code>MaxLikUpdate</code></a></p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/AbstractUpdate.jl#L23-L53" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.ZeroDriftOrnsteinUhlenbeck" href="#MolecularEvolution.ZeroDriftOrnsteinUhlenbeck"><span class="jlbinding">MolecularEvolution.ZeroDriftOrnsteinUhlenbeck</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> ZeroDriftOrnsteinUhlenbeck</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    mean</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    volatility</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    reversion</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>A 1D continuous Ornstein-Uhlenbeck process with mean, volatility, and reversion.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/continuous_models/ornstein_uhlenbeck.jl#L3-L11" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="Base.:==-Union{Tuple{T}, Tuple{T, T}} where T&lt;:AbstractTreeNode" href="#Base.:==-Union{Tuple{T}, Tuple{T, T}} where T&lt;:AbstractTreeNode"><span class="jlbinding">Base.:==</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">==</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(t1, t2)</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">Defaults to pointer equality</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/AbstractTreeNode.jl#L78-L81" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.BayesUpdate-Tuple{}" href="#MolecularEvolution.BayesUpdate-Tuple{}"><span class="jlbinding">MolecularEvolution.BayesUpdate</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">BayesUpdate</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(;</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    nni </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    branchlength </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    root </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    models </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    refresh </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    branchlength_sampler</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">UnivariateSampler</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> =</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> BranchlengthSampler</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">        Normal</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">2</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">),</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">        Normal</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">-</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">),</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    ),</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    root_sampler</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">RootSample</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> =</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> StandardRootSample</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1.0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">),</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    models_sampler</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">ModelsUpdate</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> =</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> StandardModelsUpdate</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">()</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Convenience constructor for <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.StandardUpdate"><code>StandardUpdate</code></a>. The <code>nni_selection</code> is fixed to <code>softmax_sampler</code>. This constructor provides Bayesian updates by sampling from the posterior distribution.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/AbstractUpdate.jl#L66-L83" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.HB98_AAfit-Tuple{Any, Any, Any}" href="#MolecularEvolution.HB98_AAfit-Tuple{Any, Any, Any}"><span class="jlbinding">MolecularEvolution.HB98_AAfit</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> HB98_AAfit</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(alpha, nuc_matrix, AA_fitness; genetic_code </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> universal_code)</span></span></code></pre></div><p>Compute the (Halpern and Bruno 1998) codon substitution Q matrix parameterized by <code>alpha</code>, a nucleotide substitution matrix <code>nuc_matrix</code>, and direct amino acid fitness <code>AA_fitness</code> vector.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/codon_models.jl#L304-L309" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.HB98_F61-Tuple{Any, Any, Any}" href="#MolecularEvolution.HB98_F61-Tuple{Any, Any, Any}"><span class="jlbinding">MolecularEvolution.HB98_F61</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> HB98_F61</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(alpha, nuc_matrix, F61; genetic_code </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> universal_code)</span></span></code></pre></div><p>Compute the (Halpern and Bruno 1998) HB98-F61 codon substitution Q matrix parameterized by <code>alpha</code>, a nucleotide substitution matrix <code>nuc_matrix</code>, and the <code>F61</code> codon frequency estimator (see <code>?MolecularEvolution.count_F61</code>).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/codon_models.jl#L264-L269" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.HIPSTR-Tuple{Vector{FelNode}}" href="#MolecularEvolution.HIPSTR-Tuple{Vector{FelNode}}"><span class="jlbinding">MolecularEvolution.HIPSTR</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">HIPSTR</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(trees</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{FelNode}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; set_branchlengths </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> true</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Construct a Highest Independent Posterior Subtree Reconstruction (HIPSTR) tree from a collection of trees.</p><p>Returns a single FelNode representing the HIPSTR consensus tree.</p><p>If <code>set_branchlengths = true</code>, the branch length of a node in the HIPSTR tree will be set to the mean branch length of all nodes from the input trees that have the same clade. (By the same clade, we mean that the set of leaves below the node is the same.) Otherwise, the root branch length is 0.0 and the rest 1.0.</p><p>Source: <a href="https://www.biorxiv.org/content/10.1101/2024.12.08.627395v1.full.pdf" target="_blank" rel="noreferrer">https://www.biorxiv.org/content/10.1101/2024.12.08.627395v1.full.pdf</a></p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/HIPSTR.jl#L2-L13" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.MG94_F3x4-NTuple{4, Any}" href="#MolecularEvolution.MG94_F3x4-NTuple{4, Any}"><span class="jlbinding">MolecularEvolution.MG94_F3x4</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> MG94_F3x4</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(alpha, beta, nuc_matrix, F3x4; genetic_code </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> universal_code)</span></span></code></pre></div><p>Compute the (Muse and Gaut 1994) MG94-F3x4 codon substitution Q matrix parameterized by <code>alpha</code>, <code>beta</code>, a nucleotide substitution matrix <code>nuc_matrix</code>, and <code>F3x4</code> (see <code>?MolecularEvolution.count_F3x4</code>).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/codon_models.jl#L202-L207" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.MG94_F61-NTuple{4, Any}" href="#MolecularEvolution.MG94_F61-NTuple{4, Any}"><span class="jlbinding">MolecularEvolution.MG94_F61</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> MG94_F61</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(alpha, beta, nuc_matrix, F61; genetic_code </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> universal_code)</span></span></code></pre></div><p>Compute the (Muse and Gaut 1994) MG94-F61 codon substitution Q matrix parameterized by <code>alpha</code>, <code>beta</code>, a nucleotide substitution matrix <code>nuc_matrix</code>, and the <code>F61</code> codon frequency estimator (see <code>?MolecularEvolution.count_F61</code>).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/codon_models.jl#L243-L248" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.MaxLikUpdate-Tuple{}" href="#MolecularEvolution.MaxLikUpdate-Tuple{}"><span class="jlbinding">MolecularEvolution.MaxLikUpdate</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">MaxLikUpdate</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(;</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    nni </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    branchlength </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    root </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    models </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    refresh </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    branchlength_optimizer</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">UnivariateOpt</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> =</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> GoldenSectionOpt</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(),</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    root_optimizer </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> StandardRootOpt</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">),</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    models_optimizer</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">ModelUpdate</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> =</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> StandardModelUpdate</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">()</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Convenience constructor for <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.StandardUpdate"><code>StandardUpdate</code></a>. The <code>nni_selection</code> is fixed to <code>argmax</code>. This constructor provides Maximum Likelihood updates by optimizing parameters.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/AbstractUpdate.jl#L108-L122" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.SWM_prob_grid-Union{Tuple{SWMPartition{PType}}, Tuple{PType}} where PType&lt;:MultiSitePartition" href="#MolecularEvolution.SWM_prob_grid-Union{Tuple{SWMPartition{PType}}, Tuple{PType}} where PType&lt;:MultiSitePartition"><span class="jlbinding">MolecularEvolution.SWM_prob_grid</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWM_prob_grid</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(part</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMPartition{PType}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> {PType </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> MultiSitePartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">}</span></span></code></pre></div><p>Returns a matrix of probabilities for each site, for each model (in the probability domain - not logged!) as well as the log probability offsets</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/compound_models/swm.jl#L150-L154" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution._mapreduce-Union{Tuple{T}, Tuple{AbstractTreeNode, T, Any, Any}} where T&lt;:Function" href="#MolecularEvolution._mapreduce-Union{Tuple{T}, Tuple{AbstractTreeNode, T, Any, Any}} where T&lt;:Function"><span class="jlbinding">MolecularEvolution._mapreduce</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Internal function. Helper for bfs_mapreduce and dfs_mapreduce</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/base_tree_utils.jl#L1" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.allocate!-Tuple{Any, Any}" href="#MolecularEvolution.allocate!-Tuple{Any, Any}"><span class="jlbinding">MolecularEvolution.allocate!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">allocate!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree, partition_or_message)</span></span></code></pre></div><p>Allocates initial messages for all nodes in the tree, copying the passed-in message template. If passed a partition, then this will assume the message template is a vector containing just that partition.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/simple_interface.jl#L38-L43" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.backward!-Tuple{DiscretePartition, DiscretePartition, MolecularEvolution.PMatrixModel, FelNode}" href="#MolecularEvolution.backward!-Tuple{DiscretePartition, DiscretePartition, MolecularEvolution.PMatrixModel, FelNode}"><span class="jlbinding">MolecularEvolution.backward!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">backward!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(dest</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Partition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, source</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Partition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, model</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">BranchModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, node</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Propagate the source partition backwards along the branch to the destination partition, under the model. Note: You should overload this for your own BranchModel types.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/pmatrix_models/pmatrix_models.jl#L5-L10" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.bfs_mapreduce-Union{Tuple{T}, Tuple{AbstractTreeNode, T, Any}} where T&lt;:Function" href="#MolecularEvolution.bfs_mapreduce-Union{Tuple{T}, Tuple{AbstractTreeNode, T, Any}} where T&lt;:Function"><span class="jlbinding">MolecularEvolution.bfs_mapreduce</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Performs a BFS map-reduce over the tree, starting at a given node For each node, map_reduce is called as: map_reduce(curr_node::FelNode, prev_node::FelNode, aggregator) where prev_node is the previous node visited on the path from the start node to the current node It is expected to update the aggregator, and not return anything.</p><p>Not exactly conventional map-reduce, as map-reduce calls may rely on state in the aggregator added by map-reduce calls on other nodes visited earlier.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/base_tree_utils.jl#L27-L36" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.branchlength_optim!-Tuple" href="#MolecularEvolution.branchlength_optim!-Tuple"><span class="jlbinding">MolecularEvolution.branchlength_optim!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">branchlength_optim!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models;  </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">keyword arguments</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Uses golden section search, or optionally Brent&#39;s method, to optimize all branches recursively, maintaining the integrity of the messages. Requires felsenstein!() to have been run first. models can either be a single model (if the messages on the tree contain just one Partition) or an array of models, if the messages have &gt;1 Partition, or a function that takes a node, and returns a Vector{&lt;:BranchModel} if you need the models to vary from one branch to another.</p><p><strong>Keyword Arguments</strong></p><ul><li><p><code>partition_list=nothing</code>: (eg. 1:3 or [1,3,5]) lets you choose which partitions to run over (but you probably want to optimize branch lengths with all models, the default option).</p></li><li><p><code>tol=1e-5</code>: absolute tolerance for the <code>bl_optimizer</code>.</p></li><li><p><code>bl_optimizer::UnivariateModifier=GoldenSectionOpt()</code>: the algorithm used to optimize the log likelihood of a branch length. In addition to golden section search, Brent&#39;s method can be used by setting <code>bl_optimizer=BrentsMethodOpt()</code>.</p></li><li><p><code>sort_tree=false</code>: determines if a <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.lazysort!-Tuple{FelNode}"><code>lazysort!</code></a> will be performed, which can reduce the amount of temporary messages that has to be initialized.</p></li><li><p><code>traversal=Iterators.reverse</code>: a function that determines the traversal, permutes an iterable.</p></li><li><p><code>shuffle=false</code>: do a randomly shuffled traversal, overrides <code>traversal</code>.</p></li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/branchlength_optim.jl#L191-L206" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.branchlength_update!-Tuple{MolecularEvolution.UnivariateModifier, FelNode, Any}" href="#MolecularEvolution.branchlength_update!-Tuple{MolecularEvolution.UnivariateModifier, FelNode, Any}"><span class="jlbinding">MolecularEvolution.branchlength_update!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">branchlength_update!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(bl_modifier</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">UnivariateModifier</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">keyword arguments</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>A more general version of <a href="/MolecularEvolution.jl/previews/PR57/optimization#MolecularEvolution.branchlength_optim!"><code>branchlength_optim!</code></a>. Here <code>bl_modifier</code> can be either an optimizer or a sampler (or more generally, a UnivariateModifier).</p><p><strong>Keyword Arguments</strong></p><p>See <a href="/MolecularEvolution.jl/previews/PR57/optimization#MolecularEvolution.branchlength_optim!"><code>branchlength_optim!</code></a>.</p><div class="tip custom-block"><p class="custom-block-title">Note</p><p><code>bl_modifier</code> is a positional argument here, and not a keyword argument.</p></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/branchlength_optim.jl#L156-L165" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.brents_method_minimize-Tuple{Any, Real, Real, Any, Real}" href="#MolecularEvolution.brents_method_minimize-Tuple{Any, Real, Real, Any, Real}"><span class="jlbinding">MolecularEvolution.brents_method_minimize</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">brents_method_minimize</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(f, a</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, b</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, transform, t</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; ε</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">sqrt</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">eps</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">()))</span></span></code></pre></div><p>Brent&#39;s method for minimization.</p><p>Given a function f with a single local minimum in the interval (a,b), Brent&#39;s method returns an approximation of the x-value that minimizes f to an accuaracy between 2tol and 3tol, where tol is a combination of a relative and an absolute tolerance, tol := ε|x| + t. ε should be no smaller <code>2*eps</code>, and preferably not much less than <code>sqrt(eps)</code>, which is also the default value. eps is defined here as the machine epsilon in double precision. t should be positive.</p><p>The method combines the stability of a Golden Section Search and the superlinear convergence Successive Parabolic Interpolation has under certain conditions. The method never converges much slower than a Fibonacci search and for a sufficiently well-behaved f, convergence can be exptected to be superlinear, with an order that&#39;s usually atleast 1.3247...</p><p><strong>Examples</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">julia</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> f</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(x) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> exp</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">-</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">x) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">-</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> cos</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(x)</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">f (generic </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> with </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> method)</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">julia</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> m </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> brents_method_minimize</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(f, </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">-</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">2</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, identity, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1e-7</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">0.5885327257940255</span></span></code></pre></div><p>From: Richard P. Brent, &quot;Algorithms for Minimization without Derivatives&quot; (1973). Chapter 5.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/simple_optim.jl#L110-L139" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.cascading_max_state_dict-Tuple{FelNode, Any}" href="#MolecularEvolution.cascading_max_state_dict-Tuple{FelNode, Any}"><span class="jlbinding">MolecularEvolution.cascading_max_state_dict</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">cascading_max_state_dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, model; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message), node_message_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> Dict{FelNode,Vector{&lt;:Partition}}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">())</span></span></code></pre></div><p>Takes in a tree and a model (which can be a single model, an array of models, or a function that maps FelNode-&gt;Array{&lt;:BranchModel}), and returns a dictionary mapping nodes to their inferred ancestors under the following scheme: the state that maximizes the marginal likelihood is selected at the root, and then, for each node, the maximum likelihood state is selected conditioned on the maximized state of the parent node and the observations of all descendents. A subset of partitions can be specified by partition_list, and a dictionary can be passed in to avoid re-allocating memory, in case you&#39;re running this over and over.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/ancestors.jl#L175-L182" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.char_proportions-Tuple{Any, String}" href="#MolecularEvolution.char_proportions-Tuple{Any, String}"><span class="jlbinding">MolecularEvolution.char_proportions</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">char_proportions</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(seqs, alphabet</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">String</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Takes a vector of sequences and returns a vector of the proportion of each character across all sequences. An example <code>alphabet</code> argument is <code>MolecularEvolution.AAstring</code>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/utils/matrix_helpers.jl#L87-L92" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.collect_clades!-Tuple{FelNode, Dict{Tuple{UInt64, UInt64}, MolecularEvolution.CladeStats}, Dict{String, Int64}}" href="#MolecularEvolution.collect_clades!-Tuple{FelNode, Dict{Tuple{UInt64, UInt64}, MolecularEvolution.CladeStats}, Dict{String, Int64}}"><span class="jlbinding">MolecularEvolution.collect_clades!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Recursively collect clades from a tree.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/HIPSTR.jl#L183-L185" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.collect_leaf_dists-Tuple{Vector{&lt;:AbstractTreeNode}}" href="#MolecularEvolution.collect_leaf_dists-Tuple{Vector{&lt;:AbstractTreeNode}}"><span class="jlbinding">MolecularEvolution.collect_leaf_dists</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">collect_leaf_dists</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(trees</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:AbstractTreeNode}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">Returns a list of distance matrices containing the distance between the leaf nodes, which can be used to assess mixing.</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/bayes/sampling.jl#L137-L141" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.colored_seq_draw-Tuple{Any, Any, AbstractString}" href="#MolecularEvolution.colored_seq_draw-Tuple{Any, Any, AbstractString}"><span class="jlbinding">MolecularEvolution.colored_seq_draw</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">colored_seq_draw</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(x, y, str</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">AbstractString</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; color_dict</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(), font_size</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">8</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">pt, posx</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">hcenter, posy</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">vcenter)</span></span></code></pre></div><p>Draw an arbitrary sequence. <code>color_dict</code> gives a mapping from characters to colors (default black). Default options for nucleotide colorings and amino acid colorings are given in the constants <code>NUC_COLORS</code> and <code>AA_COLORS</code>. This can be used along with <code>compose_dict</code> for drawing sequences at nodes in a tree (see <code>tree_draw</code>). Returns a Compose container.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L696-L704" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.combine!-Tuple{DiscretePartition, DiscretePartition}" href="#MolecularEvolution.combine!-Tuple{DiscretePartition, DiscretePartition}"><span class="jlbinding">MolecularEvolution.combine!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">combine!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(dest</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">P</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, src</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">P</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> P</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Partition</span></span></code></pre></div><p>Combines evidence from two partitions of the same type, storing the result in dest. Note: You should overload this for your own Partititon types.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/discrete_partitions.jl#L185-L190" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.copy_tree" href="#MolecularEvolution.copy_tree"><span class="jlbinding">MolecularEvolution.copy_tree</span></a> <span class="VPBadge info jlObjectType jlFunction"><!--[-->Function<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> copy_tree</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(root</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, shallow_copy</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">Returns an untangled copy of the tree. Optionally, the flag </span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;">`shallow_copy`</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> can be used to obtain a copy of the tree with only the names and branchlengths.</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/FelNode.jl#L102-L107" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.count_F3x4-Tuple{Array{String}}" href="#MolecularEvolution.count_F3x4-Tuple{Array{String}}"><span class="jlbinding">MolecularEvolution.count_F3x4</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> count_F3x4</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(seqs</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Array{String}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Compute the F3x4 matrix from a set of sequences.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/codon_models.jl#L176-L180" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.count_F61-Tuple{Array{String}}" href="#MolecularEvolution.count_F61-Tuple{Array{String}}"><span class="jlbinding">MolecularEvolution.count_F61</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> count_F61</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(seqs</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Array{String}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; code </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> universal_code)</span></span></code></pre></div><p>Compute the F61 codon frequency vector from a set of sequences.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/codon_models.jl#L225-L229" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.deepequals-Union{Tuple{T}, Tuple{T, T}} where T&lt;:AbstractTreeNode" href="#MolecularEvolution.deepequals-Union{Tuple{T}, Tuple{T, T}} where T&lt;:AbstractTreeNode"><span class="jlbinding">MolecularEvolution.deepequals</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">deepequals</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(t1, t2)</span></span></code></pre></div><p>Checks whether two trees are equal by recursively calling this on all fields, except <code>:parent</code>, in order to prevent cycles. In order to ensure that the <code>:parent</code> field is not hiding something different on both trees, ensure that each is consistent first (see: <code>istreeconsistent</code>).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/AbstractTreeNode.jl#L88-L93" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.dfs_mapreduce-Union{Tuple{T}, Tuple{AbstractTreeNode, T, Any}} where T&lt;:Function" href="#MolecularEvolution.dfs_mapreduce-Union{Tuple{T}, Tuple{AbstractTreeNode, T, Any}} where T&lt;:Function"><span class="jlbinding">MolecularEvolution.dfs_mapreduce</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Performs a DFS map-reduce over the tree, starting at a given node See bfs_mapreduce for more details.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/base_tree_utils.jl#L46-L49" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.discrete_name_color_dict-Tuple{AbstractTreeNode, Any}" href="#MolecularEvolution.discrete_name_color_dict-Tuple{AbstractTreeNode, Any}"><span class="jlbinding">MolecularEvolution.discrete_name_color_dict</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">discrete_name_color_dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(newt</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">AbstractTreeNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,tag_func; rainbow </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, scramble </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, darken </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> true</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, col_seed </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> nothing</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Takes a tree and a tag_func, which converts the leaf label into a category (ie. there should be &lt;20 of these), and returns a color dictionary that can be used to color the leaves or bubbles.</p><p>Example tag_func: function tag_func(nam::String) return split(nam,&quot;_&quot;)[1] end</p><p>For prettier colors, but less discrimination: rainbow = true To randomize the rainbow color assignment: scramble = true col_seed is currently set to white, and excluded from the list of colors, to make them more visible.</p><p>Consider making your own version of this function to customize colors as you see fit.</p><p>Example use: num_leaves = 50 Ne_func(t) = 1*(e^-t).+5.0 newt = sim_tree(num_leaves,Ne_func,1.0,nstart = rand(1:num_leaves)); newt = ladderize(newt) tag_func(nam) = mod(sum(Int.(collect(nam))),7) dic = discrete_name_color_dict(newt,tag_func,rainbow = true); tree_draw(newt,line_width = 0.5mm,label_color_dict = dic)</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L511-L536" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.draw_example_tree-Tuple{}" href="#MolecularEvolution.draw_example_tree-Tuple{}"><span class="jlbinding">MolecularEvolution.draw_example_tree</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">draw_example_tree</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(num_leaves </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 50</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Draws a tree and shows the code that draws it.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L576-L580" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.endpoint_conditioned_sample_state_dict-Tuple{FelNode, Any}" href="#MolecularEvolution.endpoint_conditioned_sample_state_dict-Tuple{FelNode, Any}"><span class="jlbinding">MolecularEvolution.endpoint_conditioned_sample_state_dict</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">endpoint_conditioned_sample_state_dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, model; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message), node_message_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> Dict{FelNode,Vector{&lt;:Partition}}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">())</span></span></code></pre></div><p>Takes in a tree and a model (which can be a single model, an array of models, or a function that maps FelNode-&gt;Array{&lt;:BranchModel}), and draws samples under the model conditions on the leaf observations. These samples are stored in the node_message_dict, which is returned. A subset of partitions can be specified by partition_list, and a dictionary can be passed in to avoid re-allocating memory, in case you&#39;re running this over and over.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/ancestors.jl#L208-L214" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.expected_subs_per_site-Tuple{Any, Any}" href="#MolecularEvolution.expected_subs_per_site-Tuple{Any, Any}"><span class="jlbinding">MolecularEvolution.expected_subs_per_site</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">expected_subs_per_site</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(Q,mu)</span></span></code></pre></div><p>Takes a rate matrix Q and an equilibrium frequency vector, and calculates the expected number of substitutions per site.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/misc.jl#L144-L148" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.felsenstein!-Tuple{FelNode, Any}" href="#MolecularEvolution.felsenstein!-Tuple{FelNode, Any}"><span class="jlbinding">MolecularEvolution.felsenstein!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">felsenstein!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(node</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> nothing</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Should usually be called on the root of the tree. Propagates Felsenstein pass up from the tips to the root. models can either be a single model (if the messages on the tree contain just one Partition) or an array of models, if the messages have &gt;1 Partition, or a function that takes a node, and returns a Vector{&lt;:BranchModel} if you need the models to vary from one branch to another. partition_list (eg. 1:3 or [1,3,5]) lets you choose which partitions to run over.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/felsenstein.jl#L1-L8" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.felsenstein_down!-Tuple{FelNode, Any}" href="#MolecularEvolution.felsenstein_down!-Tuple{FelNode, Any}"><span class="jlbinding">MolecularEvolution.felsenstein_down!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">felsenstein_down!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(node</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message), temp_message </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> copy_message</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message))</span></span></code></pre></div><p>Should usually be called on the root of the tree. Propagates Felsenstein pass down from the root to the tips. felsenstein!() should usually be called first. models can either be a single model (if the messages on the tree contain just one Partition) or an array of models, if the messages have &gt;1 Partition, or a function that takes a node, and returns a Vector{&lt;:BranchModel} if you need the models to vary from one branch to another. partition_list (eg. 1:3 or [1,3,5]) lets you choose which partitions to run over.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/felsenstein.jl#L80-L88" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.felsenstein_roundtrip!-Tuple{FelNode, Any}" href="#MolecularEvolution.felsenstein_roundtrip!-Tuple{FelNode, Any}"><span class="jlbinding">MolecularEvolution.felsenstein_roundtrip!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">felsenstein_roundtrip!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message), temp_message </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> copy_message</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message[partition_list]))</span></span></code></pre></div><p>Should usually be called on the root of the tree. First propagates Felsenstein pass up from the tips to the root, then propagates Felsenstein pass down from the root to the tips, with the direction of time reversed (i.e. forward! = backward!). <strong>This is useful when searching for the optimal root</strong> (see <a href="/MolecularEvolution.jl/previews/PR57/optimization#MolecularEvolution.root_optim!"><code>root_optim!</code></a>). models can either be a single model (if the messages on the tree contain just one Partition) or an array of models, if the messages have &gt;1 Partition, or a function that takes a node, and returns a Vector{&lt;:BranchModel} if you need the models to vary from one branch to another. partition_list (eg. 1:3 or [1,3,5]) lets you choose which partitions to run over.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/felsenstein.jl#L170-L179" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.forward!-Tuple{DiscretePartition, DiscretePartition, MolecularEvolution.PMatrixModel, FelNode}" href="#MolecularEvolution.forward!-Tuple{DiscretePartition, DiscretePartition, MolecularEvolution.PMatrixModel, FelNode}"><span class="jlbinding">MolecularEvolution.forward!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">forward!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(dest</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Partition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, source</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Partition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, model</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">BranchModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, node</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Propagate the source partition forwards along the branch to the destination partition, under the model. Note: You should overload this for your own BranchModel types.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/pmatrix_models/pmatrix_models.jl#L22-L27" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.gappy_Q_from_symmetric_rate_matrix-Tuple{Any, Any, Any}" href="#MolecularEvolution.gappy_Q_from_symmetric_rate_matrix-Tuple{Any, Any, Any}"><span class="jlbinding">MolecularEvolution.gappy_Q_from_symmetric_rate_matrix</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">gappy_Q_from_symmetric_rate_matrix</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(sym_mat, gap_rate, eq_freqs)</span></span></code></pre></div><p>Takes a symmetric rate matrix and gap rate (governing mutations to and from gaps) and returns a gappy rate matrix. The equilibrium frequencies are multiplied on column-wise.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/utils/matrix_helpers.jl#L101-L106" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.get_highlighter_legend-Tuple{Any}" href="#MolecularEvolution.get_highlighter_legend-Tuple{Any}"><span class="jlbinding">MolecularEvolution.get_highlighter_legend</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">get_highlighter_legend</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(legend_colors)</span></span></code></pre></div><p>Returns a Compose object given an input dictionary or pairs mapping characters to colors.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L807-L811" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.get_max_depth-Tuple{Any, Real}" href="#MolecularEvolution.get_max_depth-Tuple{Any, Real}"><span class="jlbinding">MolecularEvolution.get_max_depth</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">get_max_depth</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(node,depth</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Return the maximum depth of all children starting from the indicated node.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L791-L795" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.get_phylo_tree-Tuple{FelNode}" href="#MolecularEvolution.get_phylo_tree-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.get_phylo_tree</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">get_phylo_tree</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(molev_root</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; data_function </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> (x </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">-&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> Tuple{String,Float64}[]))</span></span></code></pre></div><p>Converts a FelNode tree to a Phylo tree. The <code>data_function</code> should return a list of tuples of the form (key, value) to be added to the Phylo tree <code>data</code> Dictionary. Any key/value pairs on the FelNode <code>node_data</code> Dict will also be added to the Phylo tree.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/phylo_glue.jl#L45-L50" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.golden_section_maximize-Tuple{Any, Real, Real, Any, Real}" href="#MolecularEvolution.golden_section_maximize-Tuple{Any, Real, Real, Any, Real}"><span class="jlbinding">MolecularEvolution.golden_section_maximize</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Golden section search.</p><p>Given a function f with a single local minimum in the interval [a,b], gss returns a subset interval [c,d] that contains the minimum with d-c &lt;= tol.</p><p><strong>Examples</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">julia</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> f</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(x) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> -</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(x</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">-</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">2</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">^</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">2</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">f (generic </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> with </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> method)</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">julia</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> m </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> golden_section_maximize</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(f, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">5</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, identity, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1e-10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">2.0000000000051843</span></span></code></pre></div><p>From: <a href="https://en.wikipedia.org/wiki/Golden-section_search" target="_blank" rel="noreferrer">https://en.wikipedia.org/wiki/Golden-section_search</a></p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/simple_optim.jl#L22-L40" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.hash_clade-Tuple{BitSet}" href="#MolecularEvolution.hash_clade-Tuple{BitSet}"><span class="jlbinding">MolecularEvolution.hash_clade</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Compute a hash for a clade based on its tips.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/HIPSTR.jl#L174-L176" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.highlight_seq_draw-Tuple{Any, Any, AbstractString, Any, Any, Any}" href="#MolecularEvolution.highlight_seq_draw-Tuple{Any, Any, AbstractString, Any, Any, Any}"><span class="jlbinding">MolecularEvolution.highlight_seq_draw</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">highlight_seq_draw</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(x, y, str</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">AbstractString</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, region, basecolor, hicolor; fontsize</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">8</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">pt, posx</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">hcenter, posy</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">vcenter)</span></span></code></pre></div><p>Draw a sequence, highlighting the sites given in <code>region</code>. This can be used along with <code>compose_dict</code> for drawing sequences at nodes in a tree (see <code>tree_draw</code>). Returns a Compose container.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L641-L647" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.highlighter_tree_draw-NTuple{4, Any}" href="#MolecularEvolution.highlighter_tree_draw-NTuple{4, Any}"><span class="jlbinding">MolecularEvolution.highlighter_tree_draw</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">highlighter_tree_draw</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree, ali_seqs, seqnames, master;</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    highlighter_start </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1.1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, highlighter_width </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    coord_width </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> highlighter_start </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">+</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> highlighter_width </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">+</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    scale_length </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> nothing</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, major_breaks </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1000</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, minor_breaks </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 500</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    tree_args </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> NamedTuple[], legend_padding </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.5</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm, legend_colors </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> NUC_colors)</span></span></code></pre></div><p>Draws a combined tree and highlighter plot. The vector of seqnames must match the node names in <code>tree</code>.</p><p>kwargs:</p><ul><li><p>tree_args: kwargs to pass to <code>tree_draw()</code></p></li><li><p>legend_colors: Mapping of characters to highlighter colors (default NT_colors)</p></li><li><p>scale_length: Length of the scale bar</p></li><li><p>highlighter_start: Canvas start for the highlighter panel</p></li><li><p>highlighter_width: Canvas width for the highlighter panel</p></li><li><p>coord_width: Total width of the canvas</p></li><li><p>major_breaks: Numbered breaks for sequence axis</p></li><li><p>minor_breaks: Ticks for sequence axis</p></li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L888-L907" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.internal_message_init!-Tuple{FelNode, Partition}" href="#MolecularEvolution.internal_message_init!-Tuple{FelNode, Partition}"><span class="jlbinding">MolecularEvolution.internal_message_init!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">internal_message_init!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, partition</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Partition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">Initializes the message template </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">for</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> each node </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">in</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> the tree, </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">as</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> an array of the partition.</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/FelNode.jl#L70-L74" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.internal_message_init!-Tuple{FelNode, Vector{&lt;:Partition}}" href="#MolecularEvolution.internal_message_init!-Tuple{FelNode, Vector{&lt;:Partition}}"><span class="jlbinding">MolecularEvolution.internal_message_init!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">internal_message_init!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, empty_message</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:Partition}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">Initializes the message template </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">for</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> each node </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">in</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> the tree, allocating space </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">for</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> each partition.</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/FelNode.jl#L55-L59" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.internal_nodes-Tuple{Any}" href="#MolecularEvolution.internal_nodes-Tuple{Any}"><span class="jlbinding">MolecularEvolution.internal_nodes</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">internal_nodes</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree)</span></span></code></pre></div><p>Returns the internal nodes of the tree (including the root), as a vector.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/simple_interface.jl#L60-L64" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.istreeconsistent-Tuple{T} where T&lt;:AbstractTreeNode" href="#MolecularEvolution.istreeconsistent-Tuple{T} where T&lt;:AbstractTreeNode"><span class="jlbinding">MolecularEvolution.istreeconsistent</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">istreeconsistent</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(root)</span></span></code></pre></div><p>Checks whether the <code>:parent</code> field is set to be consistent with the <code>:child</code> field for all nodes in the subtree.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/AbstractTreeNode.jl#L61-L65" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.ladder_tree_sim-Tuple{Any}" href="#MolecularEvolution.ladder_tree_sim-Tuple{Any}"><span class="jlbinding">MolecularEvolution.ladder_tree_sim</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">ladder_tree_sim</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(ntaxa)</span></span></code></pre></div><p>Simulates a ladder-like tree, using constant population size but heterochronous sampling, under a coalescent model.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/sim_tree.jl#L154-L158" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.lazyprep!-Tuple{FelNode, Vector{&lt;:Partition}}" href="#MolecularEvolution.lazyprep!-Tuple{FelNode, Vector{&lt;:Partition}}"><span class="jlbinding">MolecularEvolution.lazyprep!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">lazyprep!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, initial_message</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:Partition}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message), direction</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">LazyDirection</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> =</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> LazyUp</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">())</span></span></code></pre></div><p>Extra, intermediate step of tree preparations between initializing messages across the tree and calling message passing algorithms with <code>LazyPartition</code>.</p><ol><li><p>Perform a <code>lazysort!</code> on tree to obtain the optimal tree for a lazy <code>felsenstein!</code> prop, or a <code>sample_down!</code>.</p></li><li><p>Fix <code>tree.parent_message</code> to an initial message.</p></li><li><p>Preallocate sufficiently many inner partitions needed for a <code>felsenstein!</code> prop, or a <code>sample_down!</code>.</p></li><li><p>Specialized preparations based on the direction of the operations (<code>forward!</code>, <code>backward!</code>). <code>LazyDown</code> or <code>LazyUp</code>.</p></li></ol><p>See also <code>LazyDown</code>, <code>LazyUp</code>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/lazy_models/lazy_partition.jl#L200-L210" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.lazysort!-Tuple{FelNode}" href="#MolecularEvolution.lazysort!-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.lazysort!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><ul><li><p>Should be run on a tree containing LazyPartitions before running <code>felsenstein!</code>. Sorts for a minimal count of active partitions during a felsenstein!</p></li><li><p>Returns the minimum length of memoryblocks (-1) required for a <code>felsenstein!</code> prop. We need a temporary memoryblock during <code>backward!</code>, hence the &#39;-1&#39;.</p></li></ul><div class="tip custom-block"><p class="custom-block-title">Note</p><p>Since felsenstein! uses a stack, we want to avoid having long node.children[1].children[1]... chains</p></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/lazy_models/lazy_partition.jl#L152-L157" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.leaf_distmat-Tuple{Any}" href="#MolecularEvolution.leaf_distmat-Tuple{Any}"><span class="jlbinding">MolecularEvolution.leaf_distmat</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">leaf_distmat</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree)</span></span></code></pre></div><p>Returns a matrix of the distances between the leaf nodes where the index on the columns and rows are sorted by the leaf names.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/bayes/sampling.jl#L150-L154" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.leaf_names-Tuple{FelNode}" href="#MolecularEvolution.leaf_names-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.leaf_names</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">leaf_names</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree)</span></span></code></pre></div><p>Returns the names of the leaves of the tree.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/simple_interface.jl#L24-L28" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.leaf_samples-Tuple{FelNode}" href="#MolecularEvolution.leaf_samples-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.leaf_samples</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">leaf_samples</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree; partition_inds </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Returns the result of <code>partition2obs</code> for each leaf of the tree. Can be used eg. after <code>sample_down!</code> is called. If using a eg. codon model, this will extract a string from the CodonPartition on each leaf. Acts upon the first partition by default, but this can be changed by setting <code>partition_inds</code>, which can also be a vector of indices, in which case the result will be a vector for each leaf.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/simple_interface.jl#L4-L11" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.leaves-Tuple{Any}" href="#MolecularEvolution.leaves-Tuple{Any}"><span class="jlbinding">MolecularEvolution.leaves</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">leaves</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree)</span></span></code></pre></div><p>Returns the leaves of the tree, as a vector.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/simple_interface.jl#L46-L50" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.linear_scale-NTuple{5, Any}" href="#MolecularEvolution.linear_scale-NTuple{5, Any}"><span class="jlbinding">MolecularEvolution.linear_scale</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">linear_scale</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(val,in_min,in_max,out_min,out_max)</span></span></code></pre></div><p>Linearly maps val which lives in [in_min,in_max] to a value in [out_min,out_max]</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/misc.jl#L199-L203" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.log_likelihood!-Tuple{FelNode, Any}" href="#MolecularEvolution.log_likelihood!-Tuple{FelNode, Any}"><span class="jlbinding">MolecularEvolution.log_likelihood!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">log_likelihood!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> nothing</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>First re-computes the upward felsenstein pass, and then computes the log likelihood of this tree. models can either be a single model (if the messages on the tree contain just one Partition) or an array of models, if the messages have &gt;1 Partition, or a function that takes a node, and returns a Vector{&lt;:BranchModel} if you need the models to vary from one branch to another. partition_list (eg. 1:3 or [1,3,5]) lets you choose which partitions to run over.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/lls.jl#L45-L52" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.log_likelihood-Tuple{FelNode, BranchModel}" href="#MolecularEvolution.log_likelihood-Tuple{FelNode, BranchModel}"><span class="jlbinding">MolecularEvolution.log_likelihood</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">log_likelihood</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> nothing</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Computed the log likelihood of this tree. Requires felsenstein!() to have been run. models can either be a single model (if the messages on the tree contain just one Partition) or an array of models, if the messages have &gt;1 Partition, or a function that takes a node, and returns a Vector{&lt;:BranchModel} if you need the models to vary from one branch to another. partition_list (eg. 1:3 or [1,3,5]) lets you choose which partitions to run over.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/lls.jl#L27-L34" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.longest_path-Tuple{FelNode}" href="#MolecularEvolution.longest_path-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.longest_path</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Returns the longest path in a tree For convenience, this is returned as two lists of form: [leaf_node, parent_node, .... root] Where the leaf_node nodes are selected to be the furthest away</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/base_tree_utils.jl#L106-L111" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.marginal_state_dict-Tuple{FelNode, Any}" href="#MolecularEvolution.marginal_state_dict-Tuple{FelNode, Any}"><span class="jlbinding">MolecularEvolution.marginal_state_dict</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">marginal_state_dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, model; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message), node_message_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> Dict{FelNode,Vector{&lt;:Partition}}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">())</span></span></code></pre></div><p>Takes in a tree and a model (which can be a single model, an array of models, or a function that maps FelNode-&gt;Array{&lt;:BranchModel}), and returns a dictionary mapping nodes to their marginal reconstructions (ie. P(state|all observations,model)). A subset of partitions can be specified by partition_list, and a dictionary can be passed in to avoid re-allocating memory, in case you&#39;re running this over and over.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/ancestors.jl#L111-L117" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.matrix_for_display-Tuple{Any, Any}" href="#MolecularEvolution.matrix_for_display-Tuple{Any, Any}"><span class="jlbinding">MolecularEvolution.matrix_for_display</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">matrix_for_display</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(Q,labels)</span></span></code></pre></div><p>Takes a numerical matrix and a vector of labels, and returns a typically mixed type matrix with the numerical values and the labels. This is to easily visualize rate matrices in eg. the REPL.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/utils/matrix_helpers.jl#L196-L201" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.metropolis_sample-Tuple{AbstractUpdate, FelNode, Any, Any}" href="#MolecularEvolution.metropolis_sample-Tuple{AbstractUpdate, FelNode, Any, Any}"><span class="jlbinding">MolecularEvolution.metropolis_sample</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">metropolis_sample</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    update!</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">AbstractUpdate</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    initial_tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    models,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    num_of_samples;</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(initial_tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message),</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    burn_in </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1000</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    sample_interval </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    collect_LLs </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    collect_models </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    midpoint_rooting </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    ladderize </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Samples tree topologies from a posterior distribution using a custom <code>update!</code> function.</p><p><strong>Arguments</strong></p><ul><li><p><code>update!</code>: A callable that takes (tree::FelNode, models) and updates <code>tree</code> and <code>models</code>. One call to <code>update!</code> corresponds to one iteration of the Metropolis algorithm.</p></li><li><p><code>initial_tree</code>: An initial tree topology with the leaves populated with data, for the likelihood calculation.</p></li><li><p><code>models</code>: A list of branch models.</p></li><li><p><code>num_of_samples</code>: The number of tree samples drawn from the posterior.</p></li><li><p><code>partition_list</code>: (eg. 1:3 or [1,3,5]) lets you choose which partitions to run over (but you probably want to sample with all partitions, the default option).</p></li><li><p><code>burn_in</code>: The number of samples discarded at the start of the Markov Chain.</p></li><li><p><code>sample_interval</code>: The distance between samples in the underlying Markov Chain (to reduce sample correlation).</p></li><li><p><code>collect_LLs</code>: Specifies if the function should return the log-likelihoods of the trees.</p></li><li><p><code>collect_models</code>: Specifies if the function should return the models.</p></li><li><p><code>midpoint_rooting</code>: Specifies whether the drawn samples should be midpoint rerooted (Important! Should only be used for time-reversible branch models starting in equilibrium).</p></li></ul><div class="tip custom-block"><p class="custom-block-title">Note</p><p>The leaves of the initial tree should be populated with data and felsenstein! should be called on the initial tree before calling this function.</p></div><p><strong>Returns</strong></p><ul><li><p><code>samples</code>: The trees drawn from the posterior. Returns shallow tree copies, which needs to be repopulated before running felsenstein! etc.</p></li><li><p><code>sample_LLs</code>: The associated log-likelihoods of the tree (optional).</p></li><li><p><code>sample_models</code>: The models drawn from the posterior (optional). The models can be collapsed into it&#39;s parameters with <code>collapse_models</code>.</p></li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/bayes/sampling.jl#L1-L37" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.metropolis_sample-Tuple{FelNode, Vector{&lt;:BranchModel}, Any}" href="#MolecularEvolution.metropolis_sample-Tuple{FelNode, Vector{&lt;:BranchModel}, Any}"><span class="jlbinding">MolecularEvolution.metropolis_sample</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">metropolis_sample</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    initial_tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:BranchModel}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    num_of_samples;</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    bl_sampler</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">UnivariateSampler</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> =</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> BranchlengthSampler</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Normal</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">2</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">), </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Normal</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">-</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">))</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    burn_in</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1000</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, </span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    sample_interval</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    collect_LLs </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    midpoint_rooting</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>A convenience method. One step of the Metropolis algorithm is performed by calling <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.nni_update!-Tuple{Function, FelNode, Any}"><code>nni_update!</code></a> with <code>softmax_sampler</code> and <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.branchlength_update!-Tuple{MolecularEvolution.UnivariateModifier, FelNode, Any}"><code>branchlength_update!</code></a> with <code>bl_sampler</code>.</p><p><strong>Additional Arguments</strong></p><ul><li><code>bl_sampler</code>: Sampler used to drawn branchlengths from the posterior.</li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/bayes/sampling.jl#L108-L124" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.metropolis_step-Tuple{Function, Any, Any}" href="#MolecularEvolution.metropolis_step-Tuple{Function, Any, Any}"><span class="jlbinding">MolecularEvolution.metropolis_step</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">metropolis_step</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(LL</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Function</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, modifier, curr_value)</span></span></code></pre></div><p>Does a standard metropolis step in an MCMC, i.e. proposes a candidate symmetrically and returns the next state in the chain, decided by the candidate being rejected or not.</p><p><strong>Interface</strong></p><p>You need a <code>MySampler &lt;: Any</code> to implement</p><ul><li><p><code>proposal(modifier::MySampler, curr_value)</code></p></li><li><p><code>log_prior(modifier::MySampler, x)</code></p></li><li><p><code>apply_decision(modifier::MySampler, accept::Bool)</code></p></li></ul><p><code>LL</code> is by default called on <code>curr_value</code> and the returned value of <code>proposal</code>. Although, it is possible to transform the current value before proposing a new value, and then take the inverse transform to match the argument <code>LL</code> expects.</p><p><strong>Extended interface</strong></p><p><strong>Hastings</strong></p><p>To allow for asymmetric proposals, you must overload</p><ul><li><code>log_proposal(modifier::MySampler, x, conditioned_on)</code></li></ul><p>which returns a constant (<code>0.0</code> in particular) by default.</p><p><strong>Transformations</strong></p><p>To make proposals in a transformed space, you overload</p><ul><li><p><code>tr(modifier::MySampler, x)</code></p></li><li><p><code>invtr(modifier::MySampler, x)</code></p></li></ul><p>which are identity transformations by default.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/simple_sample.jl#L27-L51" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.midpoint-Tuple{FelNode}" href="#MolecularEvolution.midpoint-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.midpoint</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Returns a midpoint as a node and a distance above it where the midpoint is</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/base_tree_utils.jl#L126" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.mix-Union{Tuple{SWMPartition{PType}}, Tuple{PType}} where PType&lt;:DiscretePartition" href="#MolecularEvolution.mix-Union{Tuple{SWMPartition{PType}}, Tuple{PType}} where PType&lt;:DiscretePartition"><span class="jlbinding">MolecularEvolution.mix</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">mix</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(swm_part</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMPartition{PType}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> ) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> {PType </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> MultiSitePartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">}</span></span></code></pre></div><p><code>mix</code> collapses a Site-Wise Mixture partition to a single component partition, weighted by the site-wise likelihoods for each component, and the init weights. Specifically, it takes a <code>SWMPartition{Ptype}</code> and returns a <code>PType</code>. You&#39;ll need to have this implemented for certain helper functionality if you&#39;re playing with new kinds of SWMPartitions that aren&#39;t mixtures of <code>DiscretePartitions</code>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/compound_models/swm.jl#L188-L194" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.name2node_dict-Tuple{AbstractTreeNode}" href="#MolecularEvolution.name2node_dict-Tuple{AbstractTreeNode}"><span class="jlbinding">MolecularEvolution.name2node_dict</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>name2node_dict(root)</p><p>Returns a dictionary of leaf nodes, indexed by node.name. Can be used to associate sequences with leaf nodes.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/AbstractTreeNode.jl#L540-L544" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.newick-Tuple{AbstractTreeNode}" href="#MolecularEvolution.newick-Tuple{AbstractTreeNode}"><span class="jlbinding">MolecularEvolution.newick</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">newick</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(root)</span></span></code></pre></div><p>Returns a newick string representation of the tree.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/AbstractTreeNode.jl#L626-L630" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.nni_optim!-Tuple" href="#MolecularEvolution.nni_optim!-Tuple"><span class="jlbinding">MolecularEvolution.nni_optim!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">nni_optim!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">keyword arguments</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Considers local branch swaps for all branches recursively, maintaining the integrity of the messages. Requires felsenstein!() to have been run first. models can either be a single model (if the messages on the tree contain just one Partition) or an array of models, if the messages have &gt;1 Partition, or a function that takes a node, and returns a Vector{&lt;:BranchModel} if you need the models to vary from one branch to another.</p><p><strong>Keyword Arguments</strong></p><ul><li><p><code>partition_list=nothing</code>: (eg. 1:3 or [1,3,5]) lets you choose which partitions to run over (but you probably want to optimize tree topology with all models, the default option).</p></li><li><p><code>selection_rule = x -&gt; argmax(x)</code>: a function that takes the current and proposed log likelihoods and selects a nni configuration. Note that the current log likelihood is stored at x[1].</p></li><li><p><code>sort_tree=false</code>: determines if a <a href="/MolecularEvolution.jl/previews/PR57/api#MolecularEvolution.lazysort!-Tuple{FelNode}"><code>lazysort!</code></a> will be performed, which can reduce the amount of temporary messages that has to be initialized.</p></li><li><p><code>traversal=Iterators.reverse</code>: a function that determines the traversal, permutes an iterable.</p></li><li><p><code>shuffle=false</code>: do a randomly shuffled traversal, overrides <code>traversal</code>.</p></li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/nni_optim.jl#L323-L337" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.nni_update!-Tuple{Function, FelNode, Any}" href="#MolecularEvolution.nni_update!-Tuple{Function, FelNode, Any}"><span class="jlbinding">MolecularEvolution.nni_update!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">nni_update!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(selection_rule</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Function</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">keyword arguments</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>A more verbose version of <a href="/MolecularEvolution.jl/previews/PR57/optimization#MolecularEvolution.nni_optim!"><code>nni_optim!</code></a>.</p><p><strong>Keyword Arguments</strong></p><p>See <a href="/MolecularEvolution.jl/previews/PR57/optimization#MolecularEvolution.nni_optim!"><code>nni_optim!</code></a>.</p><div class="tip custom-block"><p class="custom-block-title">Note</p><p><code>selection_rule</code> is a positional argument here, and not a keyword argument.</p></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/nni_optim.jl#L287-L296" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.node_distances-Tuple{AbstractTreeNode}" href="#MolecularEvolution.node_distances-Tuple{AbstractTreeNode}"><span class="jlbinding">MolecularEvolution.node_distances</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Compute the distance to all other nodes from a given node</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/base_tree_utils.jl#L58" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.node_names-Tuple{FelNode}" href="#MolecularEvolution.node_names-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.node_names</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">node_names</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree)</span></span></code></pre></div><p>Returns the names of the nodes of the tree.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/simple_interface.jl#L31-L35" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.node_samples-Tuple{FelNode}" href="#MolecularEvolution.node_samples-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.node_samples</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">node_samples</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree; partition_inds </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Returns the result of <code>partition2obs</code> for each node of the tree (including internal nodes, and the root). Can be used eg. after <code>sample_down!</code> is called. If using a eg. codon model, this will extract a string from the CodonPartition on each node. Acts upon the first partition by default, but this can be changed by setting <code>partition_inds</code>, which can also be a vector of indices, in which case the result will be a vector for each node.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/simple_interface.jl#L14-L21" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.nodes-Tuple{Any}" href="#MolecularEvolution.nodes-Tuple{Any}"><span class="jlbinding">MolecularEvolution.nodes</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">nodes</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree)</span></span></code></pre></div><p>Returns the nodes of the tree (including internal nodes and the root), as a vector.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/simple_interface.jl#L53-L57" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.nonreversibleQ-Tuple{Any}" href="#MolecularEvolution.nonreversibleQ-Tuple{Any}"><span class="jlbinding">MolecularEvolution.nonreversibleQ</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">nonreversibleQ</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(param_vec)</span></span></code></pre></div><p>Takes a vector of parameters and returns a nonreversible rate matrix.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/utils/matrix_helpers.jl#L151-L155" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.parent_list-Tuple{FelNode}" href="#MolecularEvolution.parent_list-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.parent_list</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Provides a list of parent nodes nodes from this node up to the root node</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/base_tree_utils.jl#L73" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.partition2obs-Tuple{DiscretePartition, String}" href="#MolecularEvolution.partition2obs-Tuple{DiscretePartition, String}"><span class="jlbinding">MolecularEvolution.partition2obs</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">partition2obs</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(part</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Partition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Extracts the most likely state from a Partition, transforming it into a convenient type. For example, a NucleotidePartition will be transformed into a nucleotide sequence of type String. Note: You should overload this for your own Partititon types.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/utils/seq_to_vec.jl#L70-L76" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.plot_multiple_trees-Tuple{Any, Any}" href="#MolecularEvolution.plot_multiple_trees-Tuple{Any, Any}"><span class="jlbinding">MolecularEvolution.plot_multiple_trees</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">plot_multiple_trees</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(trees, inf_tree; </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">keyword arguments</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Plots multiple phylogenetic trees against a reference tree, <code>inf_tree</code>. For each <strong>tree</strong> in <code>trees</code>, a linear Weighted Least Squares (WLS) problem (parameterized by the <code>weight_fn</code> keyword) is solved for the x-positions of the matching nodes between <code>inf_tree</code> and <strong>tree</strong>.</p><p><strong>Keyword Arguments</strong></p><ul><li><p><code>node_size=4</code>: the size of the nodes in the plot.</p></li><li><p><code>line_width=0.5</code>: the width of the branches from <code>trees</code>.</p></li><li><p><code>font_size=10</code>: the font size for the leaf labels.</p></li><li><p><code>margin=1.5</code>: the margin between a leaf node and its label.</p></li><li><p><code>line_alpha=0.05</code>: the transparency level of the branches from <code>trees</code>.</p></li><li><p><code>y_jitter=0.0</code>: the standard deviation of the noise in the y-coordinate.</p></li><li><p><code>weight_fn=n::FelNode -&gt; ifelse(isroot(n), 1.0, 0.0))</code>: a function that assigns a weight to a node for the WLS problem.</p></li><li><p><code>opt_scale=true</code>: whether to include a scaling parameter for the WLS problem.</p></li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/multiple_trees.jl#L74-L89" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.populate_tree!-Tuple{FelNode, Partition, Any, Any}" href="#MolecularEvolution.populate_tree!-Tuple{FelNode, Partition, Any, Any}"><span class="jlbinding">MolecularEvolution.populate_tree!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">populate_tree!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, starting_message, names, data; init_all_messages </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> true</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, tolerate_missing </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, leaf_name_transform </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> x </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">-&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> x)</span></span></code></pre></div><p>Takes a tree, and a <code>starting_message</code> (which will serve as the memory template for populating messages all over the tree). <code>starting_message</code> can be a message (ie. a vector of Partitions), but will also work with a single Partition (although the tree) will still be populated with a length-1 vector of Partitions. Further, as long as <code>obs2partition</code> is implemented for your Partition type, the leaf nodes will be populated with the data from <code>data</code>, matching the names on each leaf. When a leaf on the tree has a name that doesn&#39;t match anything in <code>names</code>, then if</p><ul><li><p><code>tolerate_missing = 0</code>, an error will be thrown</p></li><li><p><code>tolerate_missing = 1</code>, a warning will be thrown, and the message will be set to the uninformative message (requires identity!(::Partition) to be defined)</p></li><li><p><code>tolerate_missing = 2</code>, the message will be set to the uninformative message, without warnings (requires identity!(::Partition) to be defined)</p></li></ul><p>A renaming function that can eg. strip tags from the tree when matching leaf names with <code>names</code> can be passed to <code>leaf_name_transform</code></p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/misc.jl#L110-L122" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.promote_internal-Tuple{FelNode}" href="#MolecularEvolution.promote_internal-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.promote_internal</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">promote_internal</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Creates a new tree similar to the given tree, but with &#39;dummy&#39; leaf nodes (w/ zero branchlength) representing each internal node (for drawing / evenly spacing labels internal nodes).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L615-L620" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.quadratic_CI-Tuple{Function, Vector, Int64}" href="#MolecularEvolution.quadratic_CI-Tuple{Function, Vector, Int64}"><span class="jlbinding">MolecularEvolution.quadratic_CI</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">quadratic_CI</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(f</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Function</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,opt_params</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, param_ind</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; rate_conf_level </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.99</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, nudge_amount </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.01</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Takes a NEGATIVE log likelihood function (compatible with Optim.jl), a vector of maximizing parameters, an a parameter index. Returns the quadratic confidence interval.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/simple_optim.jl#L357-L362" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.quadratic_CI-Tuple{Vector, Vector}" href="#MolecularEvolution.quadratic_CI-Tuple{Vector, Vector}"><span class="jlbinding">MolecularEvolution.quadratic_CI</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">quadratic_CI</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(xvec,yvec; rate_conf_level </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.99</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Takes xvec, a vector of parameter values, and yvec, a vector of log likelihood evaluations (note: NOT the negative LLs you) might use with Optim.jl. Returns the confidence intervals computed by a quadratic approximation to the LL.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/simple_optim.jl#L333-L339" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.read_fasta-Tuple{String}" href="#MolecularEvolution.read_fasta-Tuple{String}"><span class="jlbinding">MolecularEvolution.read_fasta</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">read_fasta</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(filepath</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">String</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Reads in a fasta file and returns a tuple of (seqnames, seqs).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/fasta_io.jl#L4-L8" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.read_newick_tree-Tuple{String}" href="#MolecularEvolution.read_newick_tree-Tuple{String}"><span class="jlbinding">MolecularEvolution.read_newick_tree</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>read_newick_tree(treefile)</p><p>Reads in a tree from a file, of type FelNode</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/misc.jl#L256-L260" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.refresh!-Tuple{FelNode, Any}" href="#MolecularEvolution.refresh!-Tuple{FelNode, Any}"><span class="jlbinding">MolecularEvolution.refresh!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">refresh!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message))</span></span></code></pre></div><p>Run <code>felsenstein!</code> and <code>felsenstein_down!</code> on <code>tree</code> with <code>models</code> to refresh messages.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/AbstractUpdate.jl#L144-L148" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.reversibleQ-Tuple{Any, Any}" href="#MolecularEvolution.reversibleQ-Tuple{Any, Any}"><span class="jlbinding">MolecularEvolution.reversibleQ</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">reversibleQ</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(param_vec,eq_freqs)</span></span></code></pre></div><p>Takes a vector of parameters and equilibrium frequencies and returns a reversible rate matrix. The parameters are the upper triangle of the rate matrix, with the diagonal elements omitted, and the equilibrium frequencies are multiplied column-wise.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/utils/matrix_helpers.jl#L122-L128" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.root2tip_distances-Tuple{AbstractTreeNode}" href="#MolecularEvolution.root2tip_distances-Tuple{AbstractTreeNode}"><span class="jlbinding">MolecularEvolution.root2tip_distances</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">root2tips</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(root</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">AbstractTreeNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Returns a vector of root-to-tip distances, and a node-to-index dictionary. Be aware that this dictionary will break when any of the node content (ie. anything on the tree) changes.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/AbstractTreeNode.jl#L773-L778" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.root_optim!-Tuple{FelNode, Any}" href="#MolecularEvolution.root_optim!-Tuple{FelNode, Any}"><span class="jlbinding">MolecularEvolution.root_optim!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">root_optim!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">keyword arguments</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Optimizes the root position and root state of a tree. Returns the new, optimal root node. models can either be a single model (if the messages on the tree contain just one Partition) or an array of models, if the messages have &gt;1 Partition, or a function that takes a node, and returns a Vector{&lt;:BranchModel} if you need the models to vary from one branch to another.</p><p><strong>Keyword Arguments</strong></p><ul><li><p><code>partition_list=1:length(tree.message)</code>: (eg. 1:3 or [1,3,5]) lets you choose which partitions to run over (but you probably want to optimize root position and root state with all models, the default option).</p></li><li><p><code>root_LL!=default_root_LL_wrapper(tree.parent_message[partition_list])</code>: a function that takes a message and returns a (optimal) parent message and LL (log likelihood). The default option uses the constant <code>tree.parent_message[partition_list]</code> as parent message for all root-candidates.</p></li><li><p><code>K=10</code>: the number of equidistant root-candidate points along a branch. (only to be used in the frequentist framework!?)</p></li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/root_optim.jl#L108-L119" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.root_update!-Tuple{RootUpdate, FelNode, Any}" href="#MolecularEvolution.root_update!-Tuple{RootUpdate, FelNode, Any}"><span class="jlbinding">MolecularEvolution.root_update!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">root_update!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(root_update</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">RootUpdate</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message))</span></span></code></pre></div><p>A more general version of <a href="/MolecularEvolution.jl/previews/PR57/optimization#MolecularEvolution.root_optim!"><code>root_optim!</code></a>. Here <code>root_update</code> can be either an optimization or a sampling (or more generally, a RootUpdate).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/root_optim.jl#L88-L92" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.sample_down!-Tuple{FelNode, Any, Any}" href="#MolecularEvolution.sample_down!-Tuple{FelNode, Any, Any}"><span class="jlbinding">MolecularEvolution.sample_down!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>sample_down!(root::FelNode,models,partition_list)</p><p>Generates samples under the model. The root.parent_message is taken as the starting distribution, and node.message contains the sampled messages. models can either be a single model (if the messages on the tree contain just one Partition) or an array of models, if the messages have &gt;1 Partition, or a function that takes a node, and returns a Vector{&lt;:BranchModel} if you need the models to vary from one branch to another. partition_list (eg. 1:3 or [1,3,5]) lets you choose which partitions to run over.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/generative.jl#L14-L21" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.sample_from_message!-Tuple{Vector{&lt;:Partition}}" href="#MolecularEvolution.sample_from_message!-Tuple{Vector{&lt;:Partition}}"><span class="jlbinding">MolecularEvolution.sample_from_message!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">sample_from_message!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(message</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:Partition}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>#Replaces an uncertain message with a sample from the distribution represented by each partition.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/generative.jl#L5-L9" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.savefig_tweakSVG-Tuple{Any, Context}" href="#MolecularEvolution.savefig_tweakSVG-Tuple{Any, Context}"><span class="jlbinding">MolecularEvolution.savefig_tweakSVG</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">savefig_tweakSVG</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(fname, plot</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Context</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; width </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm, height </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm, linecap_round </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> true</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, white_background </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> true</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Saves a figure created using the <code>Compose</code> approach, but tweaks the SVG after export.</p><p>eg. <code>savefig_tweakSVG(&quot;export.svg&quot;,pl)</code></p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L1012-L1018" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.savefig_tweakSVG-Tuple{Any, Plots.Plot}" href="#MolecularEvolution.savefig_tweakSVG-Tuple{Any, Plots.Plot}"><span class="jlbinding">MolecularEvolution.savefig_tweakSVG</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">savefig_tweakSVG</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(fname, plot</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Plots.Plot</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; hack_bounding_box </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> true</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, new_viewbox </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> nothing</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, linecap_round </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> true</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Note: Might only work if you&#39;re using the GR backend!! Saves a figure created using the <code>Phylo</code> <code>Plots</code> recipe, but tweaks the SVG after export. <code>new_viewbox</code> needs to be an array of 4 numbers, typically starting at <code>[0 0 plot_width*4 plot_height*4]</code> but this lets you add shifts, in case the plot is getting cut off.</p><p>eg. <code>savefig_tweakSVG(&quot;export.svg&quot;,pl, new_viewbox = [-100, -100, 3000, 4500])</code></p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/phylo_glue.jl#L84-L93" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.shortest_path_between_nodes-Tuple{FelNode, FelNode}" href="#MolecularEvolution.shortest_path_between_nodes-Tuple{FelNode, FelNode}"><span class="jlbinding">MolecularEvolution.shortest_path_between_nodes</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>Shortest path between nodes, returned as two lists, each starting with one of the two nodes, and ending with the common ancestor</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/base_tree_utils.jl#L84-L87" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.sibling_inds-Tuple{AbstractTreeNode}" href="#MolecularEvolution.sibling_inds-Tuple{AbstractTreeNode}"><span class="jlbinding">MolecularEvolution.sibling_inds</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>sibling_inds(node)</p><p>Returns logical indices of the siblings in the parent&#39;s child&#39;s vector.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/AbstractTreeNode.jl#L526-L530" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.siblings-Tuple{AbstractTreeNode}" href="#MolecularEvolution.siblings-Tuple{AbstractTreeNode}"><span class="jlbinding">MolecularEvolution.siblings</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>siblings(node)</p><p>Returns a vector of siblings of node.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/AbstractTreeNode.jl#L512-L516" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.sim_tree-Tuple{Int64, Any, Any}" href="#MolecularEvolution.sim_tree-Tuple{Int64, Any, Any}"><span class="jlbinding">MolecularEvolution.sim_tree</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">sim_tree</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(add_limit</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,Ne_func,sample_rate_func; nstart </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, time </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, mutation_rate </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1.0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, T </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> Float64)</span></span></code></pre></div><p>Simulates a tree of type FelNode{T}. Allows an effective population size function (Ne_func), as well as a sample rate function (sample_rate_func), which can also just be constants.</p><p>Ne_func(t) = (sin(t/10)+1)*100.0 + 10.0 root = sim_tree(600,Ne_func,1.0) simple_tree_draw(ladderize(root))</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/sim_tree.jl#L94-L103" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.sim_tree-Tuple{}" href="#MolecularEvolution.sim_tree-Tuple{}"><span class="jlbinding">MolecularEvolution.sim_tree</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">sim_tree</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(;n </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Simulates tree with constant population size.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/sim_tree.jl#L132-L137" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.simple_radial_tree_plot-Tuple{FelNode}" href="#MolecularEvolution.simple_radial_tree_plot-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.simple_radial_tree_plot</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">simple_radial_tree_plot</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(root</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; canvas_width </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm, line_color </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;"> &quot;black&quot;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, line_width </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">mm)</span></span></code></pre></div><p>Draws a radial tree. No frills. No labels. Canvas height is automatically determined to avoid distorting the tree.</p><p>newt = better_newick_import(&quot;((A:1,B:1,C:1,D:1,E:1,F:1,G:1):1,(H:1,I:1):1);&quot;, FelNode{Float64}); simple_radial_tree_plot(newt,line_width = 0.5mm,root_angle = 7/10)</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L423-L430" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.simple_tree_draw-Tuple{FelNode}" href="#MolecularEvolution.simple_tree_draw-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.simple_tree_draw</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>img = simple_tree_draw(tree::FelNode; canvas_width = 15cm, canvas_height = 15cm, line_color = &quot;black&quot;, line_width = 0.1mm)</p><p>A line drawing of a tree with very few options.</p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">img </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> simple_tree_draw</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree)</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">img </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">|&gt;</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> SVG</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;">&quot;imgout.svg&quot;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm)</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">OR</span></span>
<span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">using</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> Cairo</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">img </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">|&gt;</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> PDF</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;">&quot;imgout.pdf&quot;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm)</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L92-L104" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.standard_tree_sim-Tuple{Any}" href="#MolecularEvolution.standard_tree_sim-Tuple{Any}"><span class="jlbinding">MolecularEvolution.standard_tree_sim</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">standard_tree_sim</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(ntaxa)</span></span></code></pre></div><p>Simulates a tree with logistic population growth, under a coalescent model.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/sim_tree.jl#L143-L147" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.total_LL-Tuple{Partition}" href="#MolecularEvolution.total_LL-Tuple{Partition}"><span class="jlbinding">MolecularEvolution.total_LL</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>total_LL(p::Partition)</p><p>If called on the root, it returns the log likelihood associated with that partition. Can be overloaded for complex partitions without straightforward site log likelihoods.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/algorithms/lls.jl#L2-L7" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.tree2distances-Tuple{AbstractTreeNode}" href="#MolecularEvolution.tree2distances-Tuple{AbstractTreeNode}"><span class="jlbinding">MolecularEvolution.tree2distances</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">tree2distances</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(root</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">AbstractTreeNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Returns a distance matrix for all pairs of leaf nodes, and a node-to-index dictionary. Be aware that this dictionary will break when any of the node content (ie. anything on the tree) changes.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/AbstractTreeNode.jl#L736-L741" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.tree2shared_branch_lengths-Tuple{AbstractTreeNode}" href="#MolecularEvolution.tree2shared_branch_lengths-Tuple{AbstractTreeNode}"><span class="jlbinding">MolecularEvolution.tree2shared_branch_lengths</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">tree2distances</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(root</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">AbstractTreeNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Returns a distance matrix for all pairs of leaf nodes, and a node-to-index dictionary. Be aware that this dictionary will break when any of the node content (ie. anything on the tree) changes.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/core/nodes/AbstractTreeNode.jl#L791-L796" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.tree_draw-Tuple{FelNode}" href="#MolecularEvolution.tree_draw-Tuple{FelNode}"><span class="jlbinding">MolecularEvolution.tree_draw</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">tree_draw</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">;</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    canvas_width </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 15</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm, canvas_height </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 15</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    stretch_for_labels </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 2.0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, draw_labels </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> true</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    line_width </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">mm, font_size </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 4</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">pt,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    min_dot_size </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.00</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, max_dot_size </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.01</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    line_opacity </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1.0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    dot_opacity </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1.0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    name_opacity </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1.0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    horizontal </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> true</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    dot_size_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> Dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(), dot_size_default </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    dot_color_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> Dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(), dot_color_default </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;"> &quot;black&quot;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    line_color_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> Dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(), line_color_default </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;"> &quot;black&quot;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    label_color_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> Dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(), label_color_default </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;"> &quot;black&quot;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    nodelabel_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> Dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(),compose_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> Dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">()</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    )</span></span></code></pre></div><p>Draws a tree with a number of self-explanatory options. Dictionaries that map a node to a color/size are used to control per-node plotting options. <code>compose_dict</code> must be a <code>FelNode-&gt;function(x,y)</code> dictionary that returns a <code>compose()</code> struct.</p><p>Example using <code>compose_dict</code></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">str_tree </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;"> &quot;(((((tax24:0.09731668728575642,(tax22:0.08792233964843627,tax18:0.9210388482867483):0.3200367900275155):0.6948314526087965,(tax13:1.9977212308725611,(tax15:0.4290074347886068,(tax17:0.32928401808187824,(tax12:0.3860215462534818,tax16:0.2197134841232339):0.1399122681886174):0.05744611946245004):1.4686085778061146):0.20724159879522402):0.4539334554156126,tax28:0.4885576926440158):0.002162260013924424,tax26:0.9451873777301325):3.8695419798779387,((tax29:0.10062813251515536,tax27:0.27653633028085006):0.04262434258357507,(tax25:0.009345653929737636,((tax23:0.015832941547076644,(tax20:0.5550597590956172,((tax8:0.6649025646927402,tax9:0.358506423199849):0.1439516404012261,tax11:0.01995439013213013):1.155181296134081):0.17930021667907567):0.10906638146207207,((((((tax6:0.013708993438720255,tax5:0.061144001556547097):0.1395453591567641,tax3:0.4713722705245479):0.07432598428904214,tax1:0.5993347898257291):1.0588025698844894,(tax10:0.13109032492533992,(tax4:0.8517302241963356,(tax2:0.8481963081549965,tax7:0.23754095940676642):0.2394313086297733):0.43596704123297675):0.08774657269409454):0.9345533723114966,(tax14:0.7089558245245173,tax19:0.444897137240675):0.08657675809803095):0.01632062723968511,tax21:0.029535281963725537):0.49502691718938285):0.25829576024240986):0.7339777396780424):4.148878039524972):0.0&quot;</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">newt </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> gettreefromnewick</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(str_tree, FelNode)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">ladderize!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(newt)</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">compose_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> Dict</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">()</span></span>
<span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">for</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> n </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">in</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> getleaflist</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(newt)</span></span>
<span class="line"><span style="--shiki-light:#6A737D;--shiki-dark:#6A737D;">    #Replace the rand(4) with the frequencies you actually want.</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    compose_dict[n] </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> (x,y)</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">-&gt;</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">pie_chart</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(x,y,MolecularEvolution</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">sum2one</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">rand</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">4</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)),size </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.03</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">end</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">tree_draw</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(newt,draw_labels </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,line_width </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.5</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">mm, compose_dict </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> compose_dict)</span></span>
<span class="line"></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">img </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> tree_draw</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree)</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">img </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">|&gt;</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> SVG</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;">&quot;imgout.svg&quot;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm)</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">OR</span></span>
<span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">using</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> Cairo</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">img </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">|&gt;</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> PDF</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#032F62;--shiki-dark:#9ECBFF;">&quot;imgout.pdf&quot;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">cm)</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/tree_compose.jl#L137-L177" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.tree_polish!-Tuple{Any, Any}" href="#MolecularEvolution.tree_polish!-Tuple{Any, Any}"><span class="jlbinding">MolecularEvolution.tree_polish!</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><p>tree_polish!(newt, models; tol = 10^-4, verbose = 1, topology = true)</p><p>Takes a tree and a model function, and optimizes branch lengths and, optionally, topology. Returns final LL. Set <code>verbose=0</code> to suppress output. Note: This is not intended for an exhaustive tree search (which requires different heuristics), but rather to polish a tree that is already relatively close to the optimum.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/misc.jl#L160-L165" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.unc2probvec-Tuple{Any}" href="#MolecularEvolution.unc2probvec-Tuple{Any}"><span class="jlbinding">MolecularEvolution.unc2probvec</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">unc2probvec</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(v)</span></span></code></pre></div><p>Takes an array of N-1 unbounded values and returns an array of N values that sums to 1. Typically useful for optimizing over categorical probability distributions.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/models/discrete_models/utils/matrix_helpers.jl#L177-L181" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.univariate_maximize-Tuple{Any, Real, Real, Any, BrentsMethodOpt, Real}" href="#MolecularEvolution.univariate_maximize-Tuple{Any, Real, Real, Any, BrentsMethodOpt, Real}"><span class="jlbinding">MolecularEvolution.univariate_maximize</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">univariate_maximize</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(f, a</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, b</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, transform, optimizer</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">BrentsMethodOpt</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, t</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; ε</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">sqrt</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(eps))</span></span></code></pre></div><p>Maximizes <code>f(x)</code> using Brent&#39;s method. See <code>?brents_method_minimize</code>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/simple_optim.jl#L198-L202" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.univariate_maximize-Tuple{Any, Real, Real, Any, GoldenSectionOpt, Real}" href="#MolecularEvolution.univariate_maximize-Tuple{Any, Real, Real, Any, GoldenSectionOpt, Real}"><span class="jlbinding">MolecularEvolution.univariate_maximize</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">univariate_maximize</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(f, a</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, b</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, transform, optimizer</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">GoldenSectionOpt</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, tol</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Real</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Maximizes <code>f(x)</code> using a Golden Section Search. See <code>?golden_section_maximize</code>.</p><p><strong>Examples</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">julia</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> f</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(x) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> -</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(x</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">-</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">2</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">^</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">2</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">f (generic </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> with </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> method)</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">julia</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&gt;</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> m </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> univariate_maximize</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(f, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">5</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, identity, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">GoldenSectionOpt</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(), </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1e-10</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">2.0000000000051843</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/simple_optim.jl#L77-L89" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.univariate_sampler-Tuple{Any, BranchlengthSampler, Any}" href="#MolecularEvolution.univariate_sampler-Tuple{Any, BranchlengthSampler, Any}"><span class="jlbinding">MolecularEvolution.univariate_sampler</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">univariate_sampler</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(LL, modifier</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">BranchlengthPeturbation</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, curr_branchlength)</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">A MCMC algorithm that draws the next sample of a Markov Chain that approximates the Posterior distrubution over the branchlengths.</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/simple_sample.jl#L18-L23" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.values_from_phylo_tree-Tuple{Any, Any}" href="#MolecularEvolution.values_from_phylo_tree-Tuple{Any, Any}"><span class="jlbinding">MolecularEvolution.values_from_phylo_tree</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">values_from_phylo_tree</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(phylo_tree, key)</span></span>
<span class="line"></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">Returns a list of values from the given key </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">in</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> the nodes of the phylo_tree, </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">in</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> an order that is somehow compatible with the order the nodes get plotted in.</span></span></code></pre></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/viz/phylo_glue.jl#L72-L77" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.weightEM-Tuple{Matrix{Float64}, Any}" href="#MolecularEvolution.weightEM-Tuple{Matrix{Float64}, Any}"><span class="jlbinding">MolecularEvolution.weightEM</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">weightEM</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(con_lik_matrix</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Array{Float64,2}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, θ; conc </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 0.0</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, iters </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 500</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Takes a conditional likelihood matrix (#categories-by-sites) and a starting frequency vector θ (length(θ) = #categories) and optimizes θ (using Expectation Maximization. Maybe.). If conc &gt; 0 then this gives something like variational bayes behavior for LDA. Maybe.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/misc.jl#L40-L45" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.write_fasta-Tuple{String, Vector{String}}" href="#MolecularEvolution.write_fasta-Tuple{String, Vector{String}}"><span class="jlbinding">MolecularEvolution.write_fasta</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">write_fasta</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(filepath</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">String</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, sequences</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{String}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; seq_names </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> nothing</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Writes a fasta file from a vector of sequences, with optional seq_names.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/fasta_io.jl#L18-L22" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.write_nexus-Tuple{String, FelNode}" href="#MolecularEvolution.write_nexus-Tuple{String, FelNode}"><span class="jlbinding">MolecularEvolution.write_nexus</span></a> <span class="VPBadge info jlObjectType jlMethod"><!--[-->Method<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">write_nexus</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(fname</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">String</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>Writes the tree as a nexus file, suitable for opening in eg. FigTree. Data in the <code>node_data</code> dictionary will be converted into annotations. Only tested for simple <code>node_data</code> formats and types.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/e2c752a0b1a5fa2ca25a6c4589b6e4bb030ee730/src/utils/misc.jl#L281-L287" target="_blank" rel="noreferrer">source</a><!--]--></span></details></div></div></main><footer class="VPDocFooter" data-v-83890dd9 data-v-4f9813fa><!--[--><!--]--><div class="edit-info" data-v-4f9813fa><div class="edit-link" data-v-4f9813fa><a class="VPLink link vp-external-link-icon no-icon edit-link-button" href="https://https://github.com/MurrellGroup/MolecularEvolution.jl/edit/main/docs/src/api.md" target="_blank" rel="noreferrer" data-v-4f9813fa><!--[--><span class="vpi-square-pen edit-link-icon" data-v-4f9813fa></span> Edit this page<!--]--></a></div><!----></div><nav class="prev-next" aria-labelledby="doc-footer-aria-label" data-v-4f9813fa><span class="visually-hidden" id="doc-footer-aria-label" data-v-4f9813fa>Pager</span><div class="pager" data-v-4f9813fa><a class="VPLink link pager-link prev" href="/MolecularEvolution.jl/previews/PR57/examples" data-v-4f9813fa><!--[--><span class="desc" data-v-4f9813fa>Previous page</span><span class="title" data-v-4f9813fa>Examples</span><!--]--></a></div><div class="pager" data-v-4f9813fa><!----></div></nav></footer><!--[--><!--]--></div></div></div><!--[--><!--]--></div></div><footer class="VPFooter has-sidebar" data-v-a9a9e638 data-v-c970a860><div class="container" data-v-c970a860><p class="message" data-v-c970a860>Made with <a href="https://luxdl.github.io/DocumenterVitepress.jl/dev/" target="_blank"><strong>DocumenterVitepress.jl</strong></a><br></p><p class="copyright" data-v-c970a860>© Copyright 2025.</p></div></footer><!--[--><!--]--></div></div>
    <script>window.__VP_HASH_MAP__=JSON.parse("{\"ancestors.md\":\"CX9IXbh7\",\"api.md\":\"B_f5Qzf5\",\"examples.md\":\"B-z5IxAV\",\"framework.md\":\"BXspTZ4r\",\"generated_update.md\":\"CMfsuHnC\",\"generated_viz.md\":\"BKluwhw0\",\"index.md\":\"BRt7Iym_\",\"intro.md\":\"BZ_9Gs9A\",\"io.md\":\"DuF8lQYA\",\"models.md\":\"bHPJku1l\",\"optimization.md\":\"DZKEV4oO\",\"simulation.md\":\"BEy3DZoa\"}");window.__VP_SITE_DATA__=JSON.parse("{\"lang\":\"en-US\",\"dir\":\"ltr\",\"title\":\"MolecularEvolution.jl\",\"description\":\"Documentation for MolecularEvolution.jl\",\"base\":\"/MolecularEvolution.jl/previews/PR57/\",\"head\":[],\"router\":{\"prefetchLinks\":true},\"appearance\":true,\"themeConfig\":{\"outline\":\"deep\",\"logo\":{\"src\":\"/logo.png\",\"width\":24,\"height\":24},\"search\":{\"provider\":\"local\",\"options\":{\"detailedView\":true}},\"nav\":[{\"text\":\"Core Concepts\",\"collapsed\":false,\"items\":[{\"text\":\"Intro\",\"link\":\"/intro\"},{\"text\":\"The MolecularEvolution.jl Framework\",\"link\":\"/framework\"},{\"text\":\"Models\",\"link\":\"/models\"}]},{\"text\":\"Methods & Algorithms\",\"collapsed\":false,\"items\":[{\"text\":\"Simulation\",\"link\":\"/simulation\"},{\"text\":\"Optimization\",\"link\":\"/optimization\"},{\"text\":\"Ancestral Reconstruction\",\"link\":\"/ancestors\"}]},{\"text\":\"Extensions & Utilities\",\"collapsed\":false,\"items\":[{\"text\":\"Input/Output\",\"link\":\"/IO\"},{\"text\":\"Visualization\",\"link\":\"/generated/viz\"},{\"text\":\"Updating a phylogenetic tree\",\"link\":\"/generated/update\"}]},{\"text\":\"Examples\",\"link\":\"/examples\"},{\"text\":\"Full API\",\"link\":\"/api\"},{\"component\":\"VersionPicker\"}],\"sidebar\":[{\"text\":\"Core Concepts\",\"collapsed\":false,\"items\":[{\"text\":\"Intro\",\"link\":\"/intro\"},{\"text\":\"The MolecularEvolution.jl Framework\",\"link\":\"/framework\"},{\"text\":\"Models\",\"link\":\"/models\"}]},{\"text\":\"Methods & Algorithms\",\"collapsed\":false,\"items\":[{\"text\":\"Simulation\",\"link\":\"/simulation\"},{\"text\":\"Optimization\",\"link\":\"/optimization\"},{\"text\":\"Ancestral Reconstruction\",\"link\":\"/ancestors\"}]},{\"text\":\"Extensions & Utilities\",\"collapsed\":false,\"items\":[{\"text\":\"Input/Output\",\"link\":\"/IO\"},{\"text\":\"Visualization\",\"link\":\"/generated/viz\"},{\"text\":\"Updating a phylogenetic tree\",\"link\":\"/generated/update\"}]},{\"text\":\"Examples\",\"link\":\"/examples\"},{\"text\":\"Full API\",\"link\":\"/api\"}],\"editLink\":{\"pattern\":\"https://https://github.com/MurrellGroup/MolecularEvolution.jl/edit/main/docs/src/:path\"},\"socialLinks\":[{\"icon\":\"github\",\"link\":\"https://github.com/MurrellGroup/MolecularEvolution.jl\"}],\"footer\":{\"message\":\"Made with <a href=\\\"https://luxdl.github.io/DocumenterVitepress.jl/dev/\\\" target=\\\"_blank\\\"><strong>DocumenterVitepress.jl</strong></a><br>\",\"copyright\":\"© Copyright 2025.\"}},\"locales\":{},\"scrollOffset\":134,\"cleanUrls\":true}");</script>
    
  </body>
</html>