<!DOCTYPE html>
<html lang="en-US" dir="ltr">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Models | MolecularEvolution.jl</title>
    <meta name="description" content="Documentation for MolecularEvolution.jl">
    <meta name="generator" content="VitePress v1.6.3">
    <link rel="preload stylesheet" href="/MolecularEvolution.jl/dev/assets/style.DiscInco.css" as="style">
    <link rel="preload stylesheet" href="/MolecularEvolution.jl/dev/vp-icons.css" as="style">
    
    <script type="module" src="/MolecularEvolution.jl/dev/assets/app.BvZhcMbI.js"></script>
    <link rel="preload" href="/MolecularEvolution.jl/dev/assets/inter-roman-latin.Di8DUHzh.woff2" as="font" type="font/woff2" crossorigin="">
    <link rel="modulepreload" href="/MolecularEvolution.jl/dev/assets/chunks/theme.vIHzkSDx.js">
    <link rel="modulepreload" href="/MolecularEvolution.jl/dev/assets/chunks/framework.B_iG_fnF.js">
    <link rel="modulepreload" href="/MolecularEvolution.jl/dev/assets/models.md.BF56sNum.lean.js">
    <link rel="icon" href="/MolecularEvolution.jl/dev/favicon.ico">
    <script src="/MolecularEvolution.jl/versions.js"></script>
    <script src="/MolecularEvolution.jl/dev/siteinfo.js"></script>
    <script id="check-dark-mode">(()=>{const e=localStorage.getItem("vitepress-theme-appearance")||"auto",a=window.matchMedia("(prefers-color-scheme: dark)").matches;(!e||e==="auto"?a:e==="dark")&&document.documentElement.classList.add("dark")})();</script>
    <script id="check-mac-os">document.documentElement.classList.toggle("mac",/Mac|iPhone|iPod|iPad/i.test(navigator.platform));</script>
  </head>
  <body>
    <div id="app"><div class="Layout" data-v-a9a9e638><!--[--><!--]--><!--[--><span tabindex="-1" data-v-492508fc></span><a href="#VPContent" class="VPSkipLink visually-hidden" data-v-492508fc>Skip to content</a><!--]--><!----><header class="VPNav" data-v-a9a9e638 data-v-f1e365da><div class="VPNavBar" data-v-f1e365da data-v-822684d1><div class="wrapper" data-v-822684d1><div class="container" data-v-822684d1><div class="title" data-v-822684d1><div class="VPNavBarTitle has-sidebar" data-v-822684d1 data-v-0f4f798b><a class="title" href="/MolecularEvolution.jl/dev/" data-v-0f4f798b><!--[--><!--]--><!--[--><img class="VPImage logo" src="/MolecularEvolution.jl/dev/logo.png" width="24" height="24" alt data-v-35a7d0b8><!--]--><span data-v-0f4f798b>MolecularEvolution.jl</span><!--[--><!--]--></a></div></div><div class="content" data-v-822684d1><div class="content-body" data-v-822684d1><!--[--><!--]--><div class="VPNavBarSearch search" data-v-822684d1><!--[--><!----><div id="local-search"><button type="button" class="DocSearch DocSearch-Button" aria-label="Search"><span class="DocSearch-Button-Container"><span class="vp-icon DocSearch-Search-Icon"></span><span class="DocSearch-Button-Placeholder">Search</span></span><span class="DocSearch-Button-Keys"><kbd class="DocSearch-Button-Key"></kbd><kbd class="DocSearch-Button-Key">K</kbd></span></button></div><!--]--></div><nav aria-labelledby="main-nav-aria-label" class="VPNavBarMenu menu" data-v-822684d1 data-v-e6d46098><span id="main-nav-aria-label" class="visually-hidden" data-v-e6d46098> Main Navigation </span><!--[--><!--[--><div class="VPFlyout VPNavBarMenuGroup active" data-v-e6d46098 data-v-04f5c5e9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" data-v-04f5c5e9><span class="text" data-v-04f5c5e9><!----><span data-v-04f5c5e9>Core Concepts</span><span class="vpi-chevron-down text-icon" data-v-04f5c5e9></span></span></button><div class="menu" data-v-04f5c5e9><div class="VPMenu" data-v-04f5c5e9 data-v-7dd3104a><div class="items" data-v-7dd3104a><!--[--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/dev/intro" data-v-acbfed09><!--[--><span data-v-acbfed09>Intro</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/dev/framework" data-v-acbfed09><!--[--><span data-v-acbfed09>The MolecularEvolution.jl Framework</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link active" href="/MolecularEvolution.jl/dev/models" data-v-acbfed09><!--[--><span data-v-acbfed09>Models</span><!--]--></a></div><!--]--><!--]--></div><!--[--><!--]--></div></div></div><!--]--><!--[--><div class="VPFlyout VPNavBarMenuGroup" data-v-e6d46098 data-v-04f5c5e9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" data-v-04f5c5e9><span class="text" data-v-04f5c5e9><!----><span data-v-04f5c5e9>Methods & Algorithms</span><span class="vpi-chevron-down text-icon" data-v-04f5c5e9></span></span></button><div class="menu" data-v-04f5c5e9><div class="VPMenu" data-v-04f5c5e9 data-v-7dd3104a><div class="items" data-v-7dd3104a><!--[--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/dev/simulation" data-v-acbfed09><!--[--><span data-v-acbfed09>Simulation</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/dev/optimization" data-v-acbfed09><!--[--><span data-v-acbfed09>Optimization</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/dev/ancestors" data-v-acbfed09><!--[--><span data-v-acbfed09>Ancestral Reconstruction</span><!--]--></a></div><!--]--><!--]--></div><!--[--><!--]--></div></div></div><!--]--><!--[--><div class="VPFlyout VPNavBarMenuGroup" data-v-e6d46098 data-v-04f5c5e9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" data-v-04f5c5e9><span class="text" data-v-04f5c5e9><!----><span data-v-04f5c5e9>Extensions & Utilities</span><span class="vpi-chevron-down text-icon" data-v-04f5c5e9></span></span></button><div class="menu" data-v-04f5c5e9><div class="VPMenu" data-v-04f5c5e9 data-v-7dd3104a><div class="items" data-v-7dd3104a><!--[--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/dev/IO" data-v-acbfed09><!--[--><span data-v-acbfed09>Input/Output</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/dev/generated/viz" data-v-acbfed09><!--[--><span data-v-acbfed09>Visualization</span><!--]--></a></div><!--]--><!--[--><div class="VPMenuLink" data-v-7dd3104a data-v-acbfed09><a class="VPLink link" href="/MolecularEvolution.jl/dev/generated/update" data-v-acbfed09><!--[--><span data-v-acbfed09>Updating a phylogenetic tree</span><!--]--></a></div><!--]--><!--]--></div><!--[--><!--]--></div></div></div><!--]--><!--[--><a class="VPLink link VPNavBarMenuLink" href="/MolecularEvolution.jl/dev/examples" tabindex="0" data-v-e6d46098 data-v-956ec74c><!--[--><span data-v-956ec74c>Examples</span><!--]--></a><!--]--><!--[--><a class="VPLink link VPNavBarMenuLink" href="/MolecularEvolution.jl/dev/api" tabindex="0" data-v-e6d46098 data-v-956ec74c><!--[--><span data-v-956ec74c>Full API</span><!--]--></a><!--]--><!--[--><!----><!--]--><!--]--></nav><!----><div class="VPNavBarAppearance appearance" data-v-822684d1 data-v-af096f4a><button class="VPSwitch VPSwitchAppearance" type="button" role="switch" title aria-checked="false" data-v-af096f4a data-v-e40a8bb6 data-v-4a1c76db><span class="check" data-v-4a1c76db><span class="icon" data-v-4a1c76db><!--[--><span class="vpi-sun sun" data-v-e40a8bb6></span><span class="vpi-moon moon" data-v-e40a8bb6></span><!--]--></span></span></button></div><div class="VPSocialLinks VPNavBarSocialLinks social-links" data-v-822684d1 data-v-164c457f data-v-ee7a9424><!--[--><a class="VPSocialLink no-icon" href="https://github.com/MurrellGroup/MolecularEvolution.jl" aria-label="github" target="_blank" rel="noopener" data-v-ee7a9424 data-v-d26d30cb><span class="vpi-social-github"></span></a><!--]--></div><div class="VPFlyout VPNavBarExtra extra" data-v-822684d1 data-v-925effce data-v-04f5c5e9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" aria-label="extra navigation" data-v-04f5c5e9><span class="vpi-more-horizontal icon" data-v-04f5c5e9></span></button><div class="menu" data-v-04f5c5e9><div class="VPMenu" data-v-04f5c5e9 data-v-7dd3104a><!----><!--[--><!--[--><!----><div class="group" data-v-925effce><div class="item appearance" data-v-925effce><p class="label" data-v-925effce>Appearance</p><div class="appearance-action" data-v-925effce><button class="VPSwitch VPSwitchAppearance" type="button" role="switch" title aria-checked="false" data-v-925effce data-v-e40a8bb6 data-v-4a1c76db><span class="check" data-v-4a1c76db><span class="icon" data-v-4a1c76db><!--[--><span class="vpi-sun sun" data-v-e40a8bb6></span><span class="vpi-moon moon" data-v-e40a8bb6></span><!--]--></span></span></button></div></div></div><div class="group" data-v-925effce><div class="item social-links" data-v-925effce><div class="VPSocialLinks social-links-list" data-v-925effce data-v-ee7a9424><!--[--><a class="VPSocialLink no-icon" href="https://github.com/MurrellGroup/MolecularEvolution.jl" aria-label="github" target="_blank" rel="noopener" data-v-ee7a9424 data-v-d26d30cb><span class="vpi-social-github"></span></a><!--]--></div></div></div><!--]--><!--]--></div></div></div><!--[--><!--[--><!--[--><div class="VPFlyout VPNolebaseEnhancedReadabilitiesMenu VPNolebaseEnhancedReadabilitiesMenuFlyout" aria-label="Enhanced Readability" role="menuitem" data-v-04f5c5e9><button type="button" class="button" aria-haspopup="true" aria-expanded="false" data-v-04f5c5e9><span class="text" data-v-04f5c5e9><span class="i-icon-park-outline:book-open option-icon" data-v-04f5c5e9></span><!----><span class="vpi-chevron-down text-icon" data-v-04f5c5e9></span></span></button><div class="menu" data-v-04f5c5e9><div class="VPMenu" data-v-04f5c5e9 data-v-7dd3104a><!----><!--[--><!--]--></div></div></div><!--]--><!--]--><!--]--><button type="button" class="VPNavBarHamburger hamburger" aria-label="mobile navigation" aria-expanded="false" aria-controls="VPNavScreen" data-v-822684d1 data-v-5dea55bf><span class="container" data-v-5dea55bf><span class="top" data-v-5dea55bf></span><span class="middle" data-v-5dea55bf></span><span class="bottom" data-v-5dea55bf></span></span></button></div></div></div></div><div class="divider" data-v-822684d1><div class="divider-line" data-v-822684d1></div></div></div><!----></header><div class="VPLocalNav has-sidebar empty" data-v-a9a9e638 data-v-070ab83d><div class="container" data-v-070ab83d><button class="menu" aria-expanded="false" aria-controls="VPSidebarNav" data-v-070ab83d><span class="vpi-align-left menu-icon" data-v-070ab83d></span><span class="menu-text" data-v-070ab83d>Menu</span></button><div class="VPLocalNavOutlineDropdown" style="--vp-vh:0px;" data-v-070ab83d data-v-168ddf5d><button data-v-168ddf5d>Return to top</button><!----></div></div></div><aside class="VPSidebar" data-v-a9a9e638 data-v-18756405><div class="curtain" data-v-18756405></div><nav class="nav" id="VPSidebarNav" aria-labelledby="sidebar-aria-label" tabindex="-1" data-v-18756405><span class="visually-hidden" id="sidebar-aria-label" data-v-18756405> Sidebar Navigation </span><!--[--><!--]--><!--[--><div class="no-transition group" data-v-9e426adc><section class="VPSidebarItem level-0 collapsible has-active" data-v-9e426adc data-v-a4b0d9bf><div class="item" role="button" tabindex="0" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><h2 class="text" data-v-a4b0d9bf>Core Concepts</h2><div class="caret" role="button" aria-label="toggle section" tabindex="0" data-v-a4b0d9bf><span class="vpi-chevron-right caret-icon" data-v-a4b0d9bf></span></div></div><div class="items" data-v-a4b0d9bf><!--[--><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/intro" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Intro</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/framework" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>The MolecularEvolution.jl Framework</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/models" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Models</p><!--]--></a><!----></div><!----></div><!--]--></div></section></div><div class="no-transition group" data-v-9e426adc><section class="VPSidebarItem level-0 collapsible" data-v-9e426adc data-v-a4b0d9bf><div class="item" role="button" tabindex="0" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><h2 class="text" data-v-a4b0d9bf>Methods & Algorithms</h2><div class="caret" role="button" aria-label="toggle section" tabindex="0" data-v-a4b0d9bf><span class="vpi-chevron-right caret-icon" data-v-a4b0d9bf></span></div></div><div class="items" data-v-a4b0d9bf><!--[--><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/simulation" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Simulation</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/optimization" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Optimization</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/ancestors" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Ancestral Reconstruction</p><!--]--></a><!----></div><!----></div><!--]--></div></section></div><div class="no-transition group" data-v-9e426adc><section class="VPSidebarItem level-0 collapsible" data-v-9e426adc data-v-a4b0d9bf><div class="item" role="button" tabindex="0" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><h2 class="text" data-v-a4b0d9bf>Extensions & Utilities</h2><div class="caret" role="button" aria-label="toggle section" tabindex="0" data-v-a4b0d9bf><span class="vpi-chevron-right caret-icon" data-v-a4b0d9bf></span></div></div><div class="items" data-v-a4b0d9bf><!--[--><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/IO" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Input/Output</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/generated/viz" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Visualization</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/generated/update" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Updating a phylogenetic tree</p><!--]--></a><!----></div><!----></div><!--]--></div></section></div><div class="no-transition group" data-v-9e426adc><section class="VPSidebarItem level-0" data-v-9e426adc data-v-a4b0d9bf><!----><div class="items" data-v-a4b0d9bf><!--[--><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/examples" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Examples</p><!--]--></a><!----></div><!----></div><div class="VPSidebarItem level-1 is-link" data-v-a4b0d9bf data-v-a4b0d9bf><div class="item" data-v-a4b0d9bf><div class="indicator" data-v-a4b0d9bf></div><a class="VPLink link link" href="/MolecularEvolution.jl/dev/api" data-v-a4b0d9bf><!--[--><p class="text" data-v-a4b0d9bf>Full API</p><!--]--></a><!----></div><!----></div><!--]--></div></section></div><!--]--><!--[--><!--]--></nav></aside><div class="VPContent has-sidebar" id="VPContent" data-v-a9a9e638 data-v-91765379><div class="VPDoc has-sidebar has-aside" data-v-91765379 data-v-83890dd9><!--[--><!--]--><div class="container" data-v-83890dd9><div class="aside" data-v-83890dd9><div class="aside-curtain" data-v-83890dd9></div><div class="aside-container" data-v-83890dd9><div class="aside-content" data-v-83890dd9><div class="VPDocAside" data-v-83890dd9 data-v-6d7b3c46><!--[--><!--]--><!--[--><!--]--><nav aria-labelledby="doc-outline-aria-label" class="VPDocAsideOutline" data-v-6d7b3c46 data-v-b38bf2ff><div class="content" data-v-b38bf2ff><div class="outline-marker" data-v-b38bf2ff></div><div aria-level="2" class="outline-title" id="doc-outline-aria-label" role="heading" data-v-b38bf2ff>On this page</div><ul class="VPDocOutlineItem root" data-v-b38bf2ff data-v-3f927ebe><!--[--><!--]--></ul></div></nav><!--[--><!--]--><div class="spacer" data-v-6d7b3c46></div><!--[--><!--]--><!----><!--[--><!--]--><!--[--><!--]--></div></div></div></div><div class="content" data-v-83890dd9><div class="content-container" data-v-83890dd9><!--[--><!--]--><main class="main" data-v-83890dd9><div style="position:relative;" class="vp-doc _MolecularEvolution_jl_dev_models" data-v-83890dd9><div><h1 id="Models" tabindex="-1">Models <a class="header-anchor" href="#Models" aria-label="Permalink to &quot;Models {#Models}&quot;">​</a></h1><p>We offer a catalogue of frequently used models that are already integrated with the framework and ready to be used. We maintain that if a model you&#39;d like to use is not included in the list, you can swiftly define one yourself and leverage our framework nonetheless.</p><h2 id="Discrete-state-models" tabindex="-1">Discrete state models <a class="header-anchor" href="#Discrete-state-models" aria-label="Permalink to &quot;Discrete state models {#Discrete-state-models}&quot;">​</a></h2><p>Here we account for a typical set-up for a discrete state <code>Partition</code>:</p><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.DiscretePartition-models" href="#MolecularEvolution.DiscretePartition-models"><span class="jlbinding">MolecularEvolution.DiscretePartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">abstract type</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> DiscretePartition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> MultiSitePartition</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> end</span></span></code></pre></div><p>Represents a <code>states</code>-by-<code>sites</code> matrix of probabilities. The following fields are loosely required:</p><ul><li><p><code>state</code>: A matrix of probabilities that are site-wise normalized.</p></li><li><p><code>states</code>: The number of states.</p></li><li><p><code>sites</code>: The number of sites.</p></li><li><p><code>scaling</code>: A vector of log-domain probability scaling factors, one per site.</p></li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/discrete_models/discrete_partitions.jl#L5-L13" target="_blank" rel="noreferrer">source</a><!--]--></span></details><p>And here&#39;s a list of simple concrete subtypes of <code>DiscretePartition</code>:</p><ul><li><p><a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.CodonPartition"><code>CodonPartition</code></a></p></li><li><p><a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.CustomDiscretePartition"><code>CustomDiscretePartition</code></a></p></li><li><p><a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.NucleotidePartition"><code>NucleotidePartition</code></a></p></li><li><p><a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.GappyNucleotidePartition"><code>GappyNucleotidePartition</code></a></p></li><li><p><a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.AminoAcidPartition"><code>AminoAcidPartition</code></a></p></li><li><p><a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.GappyAminoAcidPartition"><code>GappyAminoAcidPartition</code></a></p></li></ul><p>And then there are two typical <code>BranchModel</code>s that will cooperate with this <code>Partition</code>:</p><ul><li><p><a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.GeneralCTMC"><code>GeneralCTMC</code></a></p></li><li><p><a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.DiagonalizedCTMC"><code>DiagonalizedCTMC</code></a></p></li></ul><div class="tip custom-block"><p class="custom-block-title">Note</p><p>The two above can be regarded as special cases of the more general <a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.PModel"><code>PModel</code></a>, which just represents a P matrix.</p></div><p>A typical way of constructing your Q matrix in our ecosystem is by:</p><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.reversibleQ-models" href="#MolecularEvolution.reversibleQ-models"><span class="jlbinding">MolecularEvolution.reversibleQ</span></a> <span class="VPBadge info jlObjectType jlFunction"><!--[-->Function<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">reversibleQ</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(param_vec,eq_freqs)</span></span></code></pre></div><p>Takes a vector of parameters and equilibrium frequencies and returns a reversible rate matrix. The parameters are the upper triangle of the rate matrix, with the diagonal elements omitted, and the equilibrium frequencies are multiplied column-wise.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/discrete_models/utils/matrix_helpers.jl#L122-L128" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.nonreversibleQ-models" href="#MolecularEvolution.nonreversibleQ-models"><span class="jlbinding">MolecularEvolution.nonreversibleQ</span></a> <span class="VPBadge info jlObjectType jlFunction"><!--[-->Function<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">nonreversibleQ</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(param_vec)</span></span></code></pre></div><p>Takes a vector of parameters and returns a nonreversible rate matrix.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/discrete_models/utils/matrix_helpers.jl#L151-L155" target="_blank" rel="noreferrer">source</a><!--]--></span></details><h3 id="Codon-models" tabindex="-1">Codon models <a class="header-anchor" href="#Codon-models" aria-label="Permalink to &quot;Codon models {#Codon-models}&quot;">​</a></h3><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CodonPartition-models" href="#MolecularEvolution.CodonPartition-models"><span class="jlbinding">MolecularEvolution.CodonPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">mutable struct</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> CodonPartition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscretePartition</span></span></code></pre></div><p><strong>Constructors</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>CodonPartition(sites; code = universal_code)</span></span>
<span class="line"><span>CodonPartition(state, states, sites, scaling; code = universal_code)</span></span></code></pre></div><p><strong>Description</strong></p><p>A <code>DiscretePartition</code> where every state represents a codon. Can be customized to use different genetic codes.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/discrete_models/codon_models.jl#L349-L357" target="_blank" rel="noreferrer">source</a><!--]--></span></details><p>We offer constructors for the following Q matrix parameterizations:</p><ul><li><p><code>MolecularEvolution.MG94_F3x4</code> - for example usage, see <a href="/MolecularEvolution.jl/dev/examples#Example-3:-FUBAR">Example 3: FUBAR</a></p></li><li><p><code>MolecularEvolution.MG94_F61</code></p></li><li><p><code>MolecularEvolution.HB98_F61</code></p></li><li><p><code>MolecularEvolution.HB98_AAfit</code></p></li></ul><p>Use help mode, <code>?</code>, in the REPL to find out more about the above.</p><h3 id="Miscellaneous-models" tabindex="-1">Miscellaneous models <a class="header-anchor" href="#Miscellaneous-models" aria-label="Permalink to &quot;Miscellaneous models {#Miscellaneous-models}&quot;">​</a></h3><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.InterpolatedDiscreteModel-models" href="#MolecularEvolution.InterpolatedDiscreteModel-models"><span class="jlbinding">MolecularEvolution.InterpolatedDiscreteModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">InterpolatedDiscreteModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(siz</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, generator, tvec</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">InterpolatedDiscreteModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(Pvec</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Array{Float64,3}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, tvec</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p><code>generator</code> is a function that takes a time value <code>t</code> and returns a P matrix.</p><p><strong>Description</strong></p><p>Stores a number (<code>siz</code>) of P matrices, and the time values to which they correspond. For a requested t, the returned P matrix is (element-wise linearly) interpolated between it&#39;s two neighbours.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/discrete_models/interpolated_discrete_model.jl#L5-L15" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.PiQ-models" href="#MolecularEvolution.PiQ-models"><span class="jlbinding">MolecularEvolution.PiQ</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">PiQ</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(r</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">pi</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; normalize</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">PiQ</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">pi</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; normalize</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">false</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p><strong>Description</strong></p><p>The F81 substitution model, but for general dimensions. <a href="https://www.diva-portal.org/smash/get/diva2:1878793/FULLTEXT01.pdf" target="_blank" rel="noreferrer">https://www.diva-portal.org/smash/get/diva2:1878793/FULLTEXT01.pdf</a></p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/discrete_models/PiQ.jl#L4-L13" target="_blank" rel="noreferrer">source</a><!--]--></span></details><h2 id="Continuous-models" tabindex="-1">Continuous models <a class="header-anchor" href="#Continuous-models" aria-label="Permalink to &quot;Continuous models {#Continuous-models}&quot;">​</a></h2><p>The partition of interest is:</p><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.GaussianPartition-models" href="#MolecularEvolution.GaussianPartition-models"><span class="jlbinding">MolecularEvolution.GaussianPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">GaussianPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(mean</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, var</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, norm_const</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">GaussianPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(mean</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, var</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#6A737D;--shiki-dark:#6A737D;"># norm_const defaults to 0.0</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">GaussianPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">() </span><span style="--shiki-light:#6A737D;--shiki-dark:#6A737D;"># mean, var, norm_const default to 0.0, 1.0, 0.0 respectively</span></span></code></pre></div><p><strong>Description</strong></p><p>A partition representing a (not necessarily normalized) Gaussian distribution. <code>norm_const</code> is the log-domain normalization constant.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/continuous_models/gaussian_partition.jl#L3-L12" target="_blank" rel="noreferrer">source</a><!--]--></span></details><p>And then there are two <code>BranchModel</code>s which are compatible with the above partition, namely:</p><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.BrownianMotion-models" href="#MolecularEvolution.BrownianMotion-models"><span class="jlbinding">MolecularEvolution.BrownianMotion</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">BrownianMotion</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(mean_drift</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, var_drift</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>A 1D continuous Brownian motion model with mean drift and variance drift.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/continuous_models/brownian_motion.jl#L2-L6" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.ZeroDriftOrnsteinUhlenbeck-models" href="#MolecularEvolution.ZeroDriftOrnsteinUhlenbeck-models"><span class="jlbinding">MolecularEvolution.ZeroDriftOrnsteinUhlenbeck</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">function</span><span style="--shiki-light:#6F42C1;--shiki-dark:#B392F0;"> ZeroDriftOrnsteinUhlenbeck</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    mean</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    volatility</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">    reversion</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">,</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>A 1D continuous Ornstein-Uhlenbeck process with mean, volatility, and reversion.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/continuous_models/ornstein_uhlenbeck.jl#L3-L11" target="_blank" rel="noreferrer">source</a><!--]--></span></details><h2 id="Compound-models" tabindex="-1">Compound models <a class="header-anchor" href="#Compound-models" aria-label="Permalink to &quot;Compound models {#Compound-models}&quot;">​</a></h2><h3 id="Branch-wise-mixture" tabindex="-1">Branch-wise mixture <a class="header-anchor" href="#Branch-wise-mixture" aria-label="Permalink to &quot;Branch-wise mixture {#Branch-wise-mixture}&quot;">​</a></h3><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.BWMModel-models" href="#MolecularEvolution.BWMModel-models"><span class="jlbinding">MolecularEvolution.BWMModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">mutable struct</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> BWMModel{M} </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscreteStateModel</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> M </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> DiscreteStateModel</span></span></code></pre></div><p><strong>Fields</strong></p><ul><li><p><code>models::Vector{&lt;:M}</code>: A vector of models.</p></li><li><p><code>weights::Vector{Float64}</code>: A vector of weights.</p></li></ul><p><strong>Description</strong></p><p>Branch-wise mixture model.</p><div class="tip custom-block"><p class="custom-block-title">Note</p><p><code>forward!</code> and <code>backward!</code> are currently only defined for <code>M&lt;:PMatrixModel</code>.</p></div><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/compound_models/bwm.jl#L2-L13" target="_blank" rel="noreferrer">source</a><!--]--></span></details><h3 id="CAT" tabindex="-1">CAT <a class="header-anchor" href="#CAT" aria-label="Permalink to &quot;CAT {#CAT}&quot;">​</a></h3><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CATModel-models" href="#MolecularEvolution.CATModel-models"><span class="jlbinding">MolecularEvolution.CATModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CATModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:BranchModel}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p>CAT is something where you split the sites up, and assign each site to a different model (whose &quot;data&quot; gets stored in a contiguous block of memory).</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/compound_models/cat.jl#L2-L6" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CATPartition-models" href="#MolecularEvolution.CATPartition-models"><span class="jlbinding">MolecularEvolution.CATPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CATPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(part_inds</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Vector{Int}}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CATPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(part_inds</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Vector{Int}}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, parts</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{PType}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p><strong>Description</strong></p><p>A partition for the <a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.CATModel"><code>CATModel</code></a>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/compound_models/cat.jl#L15-L23" target="_blank" rel="noreferrer">source</a><!--]--></span></details><h3 id="Covarion" tabindex="-1">Covarion <a class="header-anchor" href="#Covarion" aria-label="Permalink to &quot;Covarion {#Covarion}&quot;">​</a></h3><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CovarionModel-models" href="#MolecularEvolution.CovarionModel-models"><span class="jlbinding">MolecularEvolution.CovarionModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CovarionModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:DiscreteStateModel}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, inter_transition_rates</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Matrix{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CovarionModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:DiscreteStateModel}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, inter_transition_rate</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Float64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span></code></pre></div><p><strong>Description</strong></p><p>The covarion model.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/compound_models/covarion.jl#L2-L10" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.CovarionPartition-models" href="#MolecularEvolution.CovarionPartition-models"><span class="jlbinding">MolecularEvolution.CovarionPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">CovarionPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(states,sites,models,t)</span></span></code></pre></div><p>A partition for the <a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.CovarionModel"><code>CovarionModel</code></a>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/compound_models/covarion.jl#L47-L51" target="_blank" rel="noreferrer">source</a><!--]--></span></details><h3 id="Site-wise-mixture" tabindex="-1">Site-wise mixture <a class="header-anchor" href="#Site-wise-mixture" aria-label="Permalink to &quot;Site-wise mixture {#Site-wise-mixture}&quot;">​</a></h3><p>See <a href="/MolecularEvolution.jl/dev/examples#Example-2:-GTRGamma">Example 2: GTR+Gamma</a> for example usage.</p><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.SWMModel-models" href="#MolecularEvolution.SWMModel-models"><span class="jlbinding">MolecularEvolution.SWMModel</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:BranchModel}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(model</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">M</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, rs</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> {M </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> BranchModel</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">}</span></span></code></pre></div><p><strong>Description</strong></p><p>A site-wise mixture model, for site-to-site &quot;random effects&quot; rate variation.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/compound_models/swm.jl#L17-L25" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.SWMPartition-models" href="#MolecularEvolution.SWMPartition-models"><span class="jlbinding">MolecularEvolution.SWMPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(parts</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{PType}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> {PType </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> MultiSitePartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">}</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(part</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">PType</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, n_parts</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> {PType </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> MultiSitePartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">}</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">SWMPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(parts</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{PType}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, weights</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{Float64}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, sites</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, states</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, models</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">) </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">where</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> {PType </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">&lt;:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> MultiSitePartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">}</span></span></code></pre></div><p><strong>Description</strong></p><p>A site-wise mixture partition for the <a href="/MolecularEvolution.jl/dev/api#MolecularEvolution.SWMModel"><code>SWMModel</code></a>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/compound_models/swm.jl#L46-L55" target="_blank" rel="noreferrer">source</a><!--]--></span></details><h2 id="Lazy-models" tabindex="-1">Lazy models <a class="header-anchor" href="#Lazy-models" aria-label="Permalink to &quot;Lazy models {#Lazy-models}&quot;">​</a></h2><h3 id="LazyPartition" tabindex="-1">LazyPartition <a class="header-anchor" href="#LazyPartition" aria-label="Permalink to &quot;LazyPartition {#LazyPartition}&quot;">​</a></h3><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.LazyPartition-models" href="#MolecularEvolution.LazyPartition-models"><span class="jlbinding">MolecularEvolution.LazyPartition</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructor</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>LazyPartition{PType}()</span></span></code></pre></div><p>Initialize an empty <code>LazyPartition</code> that is meant for wrapping a partition of type <code>PType</code>.</p><p><strong>Description</strong></p><p>With this data structure, you can wrap a partition of choice. The idea is that in some message passing algorithms, there is only a wave of partitions which need to actualize. For instance, a wave following a root-leaf path, or a depth-first traversal. In which case, we can be more economical with our memory consumption. With a worst case memory complexity of O(log(n)), where n is the number of nodes, functionality is provided for:</p><ul><li><p><code>log_likelihood!</code></p></li><li><p><code>felsenstein!</code></p></li><li><p><code>sample_down!</code></p></li></ul><div class="tip custom-block"><p class="custom-block-title">Note</p><p>For successive <code>felsenstein!</code> calls, we need to extract the information at the root somehow after each call. This can be done with e.g. <code>total_LL</code> or <code>site_LLs</code>.</p></div><p><strong>Further requirements</strong></p><p>Suppose you want to wrap a partition of <code>PType</code> with <code>LazyPartition</code>:</p><ul><li><p>If you&#39;re calling <code>log_likelihood!</code> and <code>felsenstein!</code>:</p><ul><li><code>obs2partition!(partition::PType, obs)</code> that transforms an observation to a partition.</li></ul></li><li><p>If you&#39;re calling <code>sample_down!</code>:</p><ul><li><code>partition2obs(partition::PType)</code> that returns the most likely state from a partition, inverts <code>obs2partition!</code>.</li></ul></li></ul><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/lazy_models/lazy_partition.jl#L2-L25" target="_blank" rel="noreferrer">source</a><!--]--></span></details><h4 id="Examples" tabindex="-1">Examples <a class="header-anchor" href="#Examples" aria-label="Permalink to &quot;Examples {#Examples}&quot;">​</a></h4><h5 id="Example-1:-Initializing-for-an-upward-pass" tabindex="-1">Example 1: Initializing for an upward pass <a class="header-anchor" href="#Example-1:-Initializing-for-an-upward-pass" aria-label="Permalink to &quot;Example 1: Initializing for an upward pass {#Example-1:-Initializing-for-an-upward-pass}&quot;">​</a></h5><p>Now, we show how to wrap the <code>CodonPartition</code>s from <a href="/MolecularEvolution.jl/dev/examples#Example-3:-FUBAR">Example 3: FUBAR</a> with <code>LazyPartition</code>:</p><p>You simply go from initializing messages like this:</p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">initial_partition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> CodonPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(seqs[</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">])</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">/</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">3</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">))</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">initial_partition</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">state .</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> eq_freqs</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">populate_tree!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree,initial_partition,seqnames,seqs)</span></span></code></pre></div><p>To this</p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">initial_partition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> CodonPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Int64</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(seqs[</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">1</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">])</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">/</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">3</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">))</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">initial_partition</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">state .</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;"> eq_freqs</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">lazy_initial_partition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> LazyPartition{CodonPartition}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">()</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">populate_tree!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree,lazy_initial_partition,seqnames,seqs)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">lazyprep!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree, initial_partition)</span></span></code></pre></div><p>By this slight modification, we go from initializing and using 554 partitions to 6 during the subsequent <code>log_likelihood!</code> and <code>felsenstein!</code> calls. There is no significant decrease in performance recorded from this switch.</p><h5 id="Example-2:-Initializing-for-a-downward-pass" tabindex="-1">Example 2: Initializing for a downward pass <a class="header-anchor" href="#Example-2:-Initializing-for-a-downward-pass" aria-label="Permalink to &quot;Example 2: Initializing for a downward pass {#Example-2:-Initializing-for-a-downward-pass}&quot;">​</a></h5><p>Now, we show how to wrap the <code>GaussianPartition</code>s from <a href="/MolecularEvolution.jl/dev/intro#Quick-example:-Likelihood-calculations-under-phylogenetic-Brownian-motion:">Quick example: Likelihood calculations under phylogenetic Brownian motion:</a> with <code>LazyPartition</code>:</p><p>You simply go from initializing messages like this:</p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">internal_message_init!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree, </span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">GaussianPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">())</span></span></code></pre></div><p>To this (technically we only add 1 LOC)</p><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">initial_partition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> GaussianPartition</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">()</span></span>
<span class="line"><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">lazy_initial_partition </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> LazyPartition{GaussianPartition}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">()</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">internal_message_init!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree, lazy_initial_partition)</span></span>
<span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">lazyprep!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree, initial_partition, direction</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">LazyDown</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(isleafnode))</span></span></code></pre></div><div class="tip custom-block"><p class="custom-block-title">Note</p><p>Now, we provided a direction for <code>lazyprep!</code>. The direction is an instance of <code>LazyDown</code>, which was initialized with the <code>isleafnode</code> function. The function <code>isleafnode</code> dictates if a node saves its sampled observation after a down pass. If you use <code>direction=LazyDown()</code>, every node saves its observation.</p></div><h4 id="Surrounding-LazyPartition" tabindex="-1">Surrounding LazyPartition <a class="header-anchor" href="#Surrounding-LazyPartition" aria-label="Permalink to &quot;Surrounding LazyPartition {#Surrounding-LazyPartition}&quot;">​</a></h4><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.lazyprep!-models" href="#MolecularEvolution.lazyprep!-models"><span class="jlbinding">MolecularEvolution.lazyprep!</span></a> <span class="VPBadge info jlObjectType jlFunction"><!--[-->Function<!--]--></span></summary><div class="language-julia vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang">julia</span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">lazyprep!</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">FelNode</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">, initial_message</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">Vector{&lt;:Partition}</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">; partition_list </span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">=</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> 1</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">:</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">length</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">(tree</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">.</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">message), direction</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;">::</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;">LazyDirection</span><span style="--shiki-light:#D73A49;--shiki-dark:#F97583;"> =</span><span style="--shiki-light:#005CC5;--shiki-dark:#79B8FF;"> LazyUp</span><span style="--shiki-light:#24292E;--shiki-dark:#E1E4E8;">())</span></span></code></pre></div><p>Extra, intermediate step of tree preparations between initializing messages across the tree and calling message passing algorithms with <code>LazyPartition</code>.</p><ol><li><p>Perform a <code>lazysort!</code> on tree to obtain the optimal tree for a lazy <code>felsenstein!</code> prop, or a <code>sample_down!</code>.</p></li><li><p>Fix <code>tree.parent_message</code> to an initial message.</p></li><li><p>Preallocate sufficiently many inner partitions needed for a <code>felsenstein!</code> prop, or a <code>sample_down!</code>.</p></li><li><p>Specialized preparations based on the direction of the operations (<code>forward!</code>, <code>backward!</code>). <code>LazyDown</code> or <code>LazyUp</code>.</p></li></ol><p>See also <code>LazyDown</code>, <code>LazyUp</code>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/lazy_models/lazy_partition.jl#L200-L210" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.LazyUp-models" href="#MolecularEvolution.LazyUp-models"><span class="jlbinding">MolecularEvolution.LazyUp</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructor</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>LazyUp()</span></span></code></pre></div><p><strong>Description</strong></p><p>Indicate that we want to do an upward pass, e.g. <code>felsenstein!</code>.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/lazy_models/lazy_partition.jl#L42-L48" target="_blank" rel="noreferrer">source</a><!--]--></span></details><details class="jldocstring custom-block" open><summary><a id="MolecularEvolution.LazyDown-models" href="#MolecularEvolution.LazyDown-models"><span class="jlbinding">MolecularEvolution.LazyDown</span></a> <span class="VPBadge info jlObjectType jlType"><!--[-->Type<!--]--></span></summary><p><strong>Constructors</strong></p><div class="language- vp-adaptive-theme"><button title="Copy Code" class="copy"></button><span class="lang"></span><pre class="shiki shiki-themes github-light github-dark vp-code" tabindex="0"><code><span class="line"><span>LazyDown(stores_obs)</span></span>
<span class="line"><span>LazyDown() = LazyDown(x::FelNode -&gt; true)</span></span></code></pre></div><p><strong>Description</strong></p><p>Indicate that we want to do a downward pass, e.g. <code>sample_down!</code>. The function passed to the constructor takes a <code>node::FelNode</code> as input and returns a <code>Bool</code> that decides if <code>node</code> stores its observations.</p><span class="VPBadge info source-link"><!--[--><a href="https://github.com/MurrellGroup/MolecularEvolution.jl/blob/243e6dc59c54630be29bf9034b3ff8da62d24e18/src/models/lazy_models/lazy_partition.jl#L51-L59" target="_blank" rel="noreferrer">source</a><!--]--></span></details></div></div></main><footer class="VPDocFooter" data-v-83890dd9 data-v-4f9813fa><!--[--><!--]--><div class="edit-info" data-v-4f9813fa><div class="edit-link" data-v-4f9813fa><a class="VPLink link vp-external-link-icon no-icon edit-link-button" href="https://https://github.com/MurrellGroup/MolecularEvolution.jl/edit/main/docs/src/models.md" target="_blank" rel="noreferrer" data-v-4f9813fa><!--[--><span class="vpi-square-pen edit-link-icon" data-v-4f9813fa></span> Edit this page<!--]--></a></div><!----></div><nav class="prev-next" aria-labelledby="doc-footer-aria-label" data-v-4f9813fa><span class="visually-hidden" id="doc-footer-aria-label" data-v-4f9813fa>Pager</span><div class="pager" data-v-4f9813fa><a class="VPLink link pager-link prev" href="/MolecularEvolution.jl/dev/framework" data-v-4f9813fa><!--[--><span class="desc" data-v-4f9813fa>Previous page</span><span class="title" data-v-4f9813fa>The MolecularEvolution.jl Framework</span><!--]--></a></div><div class="pager" data-v-4f9813fa><a class="VPLink link pager-link next" href="/MolecularEvolution.jl/dev/simulation" data-v-4f9813fa><!--[--><span class="desc" data-v-4f9813fa>Next page</span><span class="title" data-v-4f9813fa>Simulation</span><!--]--></a></div></nav></footer><!--[--><!--]--></div></div></div><!--[--><!--]--></div></div><footer class="VPFooter has-sidebar" data-v-a9a9e638 data-v-c970a860><div class="container" data-v-c970a860><p class="message" data-v-c970a860>Made with <a href="https://luxdl.github.io/DocumenterVitepress.jl/dev/" target="_blank"><strong>DocumenterVitepress.jl</strong></a><br></p><p class="copyright" data-v-c970a860>© Copyright 2025.</p></div></footer><!--[--><!--]--></div></div>
    <script>window.__VP_HASH_MAP__=JSON.parse("{\"ancestors.md\":\"Z9WX1VeD\",\"api.md\":\"SGVRY3A5\",\"examples.md\":\"iiUiWsaY\",\"framework.md\":\"DZGJfOfB\",\"generated_update.md\":\"O3u7hEpK\",\"generated_viz.md\":\"Dre5p6LU\",\"index.md\":\"DKFlX5Ao\",\"intro.md\":\"CWpuq1k3\",\"io.md\":\"D7mhKF_p\",\"models.md\":\"BF56sNum\",\"optimization.md\":\"BWOtDe3q\",\"simulation.md\":\"DfwsB_90\"}");window.__VP_SITE_DATA__=JSON.parse("{\"lang\":\"en-US\",\"dir\":\"ltr\",\"title\":\"MolecularEvolution.jl\",\"description\":\"Documentation for MolecularEvolution.jl\",\"base\":\"/MolecularEvolution.jl/dev/\",\"head\":[],\"router\":{\"prefetchLinks\":true},\"appearance\":true,\"themeConfig\":{\"outline\":\"deep\",\"logo\":{\"src\":\"/logo.png\",\"width\":24,\"height\":24},\"search\":{\"provider\":\"local\",\"options\":{\"detailedView\":true}},\"nav\":[{\"text\":\"Core Concepts\",\"collapsed\":false,\"items\":[{\"text\":\"Intro\",\"link\":\"/intro\"},{\"text\":\"The MolecularEvolution.jl Framework\",\"link\":\"/framework\"},{\"text\":\"Models\",\"link\":\"/models\"}]},{\"text\":\"Methods & Algorithms\",\"collapsed\":false,\"items\":[{\"text\":\"Simulation\",\"link\":\"/simulation\"},{\"text\":\"Optimization\",\"link\":\"/optimization\"},{\"text\":\"Ancestral Reconstruction\",\"link\":\"/ancestors\"}]},{\"text\":\"Extensions & Utilities\",\"collapsed\":false,\"items\":[{\"text\":\"Input/Output\",\"link\":\"/IO\"},{\"text\":\"Visualization\",\"link\":\"/generated/viz\"},{\"text\":\"Updating a phylogenetic tree\",\"link\":\"/generated/update\"}]},{\"text\":\"Examples\",\"link\":\"/examples\"},{\"text\":\"Full API\",\"link\":\"/api\"},{\"component\":\"VersionPicker\"}],\"sidebar\":[{\"text\":\"Core Concepts\",\"collapsed\":false,\"items\":[{\"text\":\"Intro\",\"link\":\"/intro\"},{\"text\":\"The MolecularEvolution.jl Framework\",\"link\":\"/framework\"},{\"text\":\"Models\",\"link\":\"/models\"}]},{\"text\":\"Methods & Algorithms\",\"collapsed\":false,\"items\":[{\"text\":\"Simulation\",\"link\":\"/simulation\"},{\"text\":\"Optimization\",\"link\":\"/optimization\"},{\"text\":\"Ancestral Reconstruction\",\"link\":\"/ancestors\"}]},{\"text\":\"Extensions & Utilities\",\"collapsed\":false,\"items\":[{\"text\":\"Input/Output\",\"link\":\"/IO\"},{\"text\":\"Visualization\",\"link\":\"/generated/viz\"},{\"text\":\"Updating a phylogenetic tree\",\"link\":\"/generated/update\"}]},{\"text\":\"Examples\",\"link\":\"/examples\"},{\"text\":\"Full API\",\"link\":\"/api\"}],\"editLink\":{\"pattern\":\"https://https://github.com/MurrellGroup/MolecularEvolution.jl/edit/main/docs/src/:path\"},\"socialLinks\":[{\"icon\":\"github\",\"link\":\"https://github.com/MurrellGroup/MolecularEvolution.jl\"}],\"footer\":{\"message\":\"Made with <a href=\\\"https://luxdl.github.io/DocumenterVitepress.jl/dev/\\\" target=\\\"_blank\\\"><strong>DocumenterVitepress.jl</strong></a><br>\",\"copyright\":\"© Copyright 2025.\"}},\"locales\":{},\"scrollOffset\":134,\"cleanUrls\":true}");</script>
    
  </body>
</html>